// aletheia/lib/aletheia/parser.js
/* parser generated by jison 0.4.13 */
/*
  Returns a Parser object of the following structure:

  Parser: {
    yy: {}
  }

  Parser.prototype: {
    yy: {},
    trace: function(),
    symbols_: {associative list: name ==> number},
    terminals_: {associative list: number ==> name},
    productions_: [...],
    performAction: function anonymous(yytext, yyleng, yylineno, yy, yystate, $$, _$),
    table: [...],
    defaultActions: {...},
    parseError: function(str, hash),
    parse: function(input),

    lexer: {
        EOF: 1,
        parseError: function(str, hash),
        setInput: function(input),
        input: function(),
        unput: function(str),
        more: function(),
        less: function(n),
        pastInput: function(),
        upcomingInput: function(),
        showPosition: function(),
        test_match: function(regex_match_array, rule_index),
        next: function(),
        lex: function(),
        begin: function(condition),
        popState: function(),
        _currentRules: function(),
        topState: function(),
        pushState: function(condition),

        options: {
            ranges: boolean           (optional: true ==> token location info will include a .range[] member)
            flex: boolean             (optional: true ==> flex-like lexing behaviour where the rules are tested exhaustively to find the longest match)
            backtrack_lexer: boolean  (optional: true ==> lexer regexes are tested in order and for each matching regex the action code is invoked; the lexer terminates the scan when a token is returned by the action code)
        },

        performAction: function(yy, yy_, $avoiding_name_collisions, YY_START),
        rules: [...],
        conditions: {associative list: name ==> set},
    }
  }


  token location info (@$, _$, etc.): {
    first_line: n,
    last_line: n,
    first_column: n,
    last_column: n,
    range: [start_number, end_number]       (where the numbers are indexes into the input string, regular zero-based)
  }


  the parseError function receives a 'hash' object with these members for lexer and parser errors: {
    text:        (matched text)
    token:       (the produced terminal token, if any)
    line:        (yylineno)
  }
  while parser (grammar) errors will also provide these members, i.e. parser errors deliver a superset of attributes: {
    loc:         (yylloc)
    expected:    (string describing the set of expected tokens)
    recoverable: (boolean: TRUE when the parser has a error recovery rule available for this particular error)
  }
*/
var parser = (function(){
var parser = {trace: function trace() { },
yy: {},
symbols_: {"error":2,"program":3,"statementList":4,"EOF":5,"statementListBody":6,"separator":7,"statement":8,"NEWLINE":9,"lvalue":10,"=":11,"expression":12,"unitList":13,"additive":14,"booleanOp":15,"arrowExpr":16,"comparison":17,"booleanAndOp":18,"booleanOrOp":19,"AND":20,"unitExpression":21,"OR":22,"SIGN":23,"(":24,")":25,"function":26,"literal":27,"singleUnitList":28,"JAVASCRIPT":29,"IDENTIFIER":30,"TYPEDELIM":31,"?":32,"tableaccess":33,"->":34,"DOT":35,"@":36,"tableNameLiteral":37,"table":38,"REGEX":39,"NUMBER":40,"NEG_NUMBER":41,"STRING":42,"{":43,"}":44,":":45,"fieldList":46,"fieldListBody":47,"field":48,",":49,"[":50,"]":51,"|":52,"multiplicative":53,"+":54,"-":55,"negative":56,"*":57,"/":58,"%":59,"$accept":0,"$end":1},
terminals_: {2:"error",5:"EOF",9:"NEWLINE",11:"=",20:"AND",22:"OR",23:"SIGN",24:"(",25:")",29:"JAVASCRIPT",30:"IDENTIFIER",31:"TYPEDELIM",32:"?",34:"->",35:"DOT",36:"@",39:"REGEX",40:"NUMBER",41:"NEG_NUMBER",42:"STRING",43:"{",44:"}",45:":",49:",",50:"[",51:"]",52:"|",54:"+",55:"-",57:"*",58:"/",59:"%"},
productions_: [0,[3,2],[3,1],[4,1],[4,2],[4,2],[4,3],[6,1],[6,3],[7,1],[7,2],[8,3],[8,3],[8,1],[12,1],[12,1],[12,1],[12,1],[15,1],[15,1],[15,1],[18,3],[18,3],[18,3],[18,3],[18,3],[18,3],[19,3],[19,3],[19,3],[19,3],[19,3],[19,3],[17,3],[21,3],[21,1],[21,1],[21,1],[21,1],[21,1],[10,1],[10,3],[10,3],[10,1],[13,2],[13,2],[28,3],[16,3],[16,3],[33,3],[33,3],[27,1],[27,1],[27,1],[37,1],[37,1],[37,1],[38,2],[38,3],[38,3],[38,3],[46,1],[46,2],[46,2],[46,3],[47,1],[47,3],[47,3],[47,4],[48,1],[48,3],[48,3],[48,3],[48,3],[26,3],[26,4],[26,5],[26,5],[26,5],[26,6],[26,7],[26,7],[14,1],[14,3],[14,3],[53,1],[53,3],[53,3],[53,3],[56,2],[56,1]],
performAction: function anonymous(yytext, yyleng, yylineno, yy, yystate /* action[1] */, $$ /* vstack */, _$ /* lstack */) {
/* this == yyval */

var $0 = $$.length - 1;
switch (yystate) {
case 1:return $$[$0-1];
break;
case 2:return [];
break;
case 3:this.$ = $$[$0];
break;
case 4:this.$ = $$[$0-1];
break;
case 5:this.$ = $$[$0];
break;
case 6:this.$ = $$[$0-1];
break;
case 7:this.$ = [$$[$0]];
break;
case 8:this.$ = $$[$0-2]; $$[$0-2].push($$[$0]);
break;
case 9:
break;
case 10:
break;
case 11:this.$ = yy.Declaration($$[$0-2], $$[$0], _$[$0-1]);
break;
case 12:this.$ = yy.Assignment($$[$0-2], $$[$0], _$[$0-1]);
break;
case 13:this.$ = $$[$0];
break;
case 14:this.$ = $$[$0];
break;
case 15:this.$ = $$[$0];
break;
case 16:this.$ = $$[$0]
break;
case 17:this.$ = $$[$0]
break;
case 18:this.$ = $$[$0]
break;
case 19:this.$ = $$[$0]
break;
case 20:this.$ = $$[$0]
break;
case 21:this.$ = yy.Operation($$[$0-2], $$[$0-1], $$[$0]);
break;
case 22:this.$ = yy.Operation($$[$0-2], $$[$0-1], $$[$0]);
break;
case 23:this.$ = yy.Operation($$[$0-2], $$[$0-1], $$[$0]);
break;
case 24:this.$ = yy.Operation($$[$0-2], $$[$0-1], $$[$0]);
break;
case 25:this.$ = yy.Operation($$[$0-2], $$[$0-1], $$[$0]);
break;
case 26:this.$ = yy.Operation($$[$0-2], $$[$0-1], $$[$0]);
break;
case 27:this.$ = yy.Operation($$[$0-2], $$[$0-1], $$[$0]);
break;
case 28:this.$ = yy.Operation($$[$0-2], $$[$0-1], $$[$0]);
break;
case 29:this.$ = yy.Operation($$[$0-2], $$[$0-1], $$[$0]);
break;
case 30:this.$ = yy.Operation($$[$0-2], $$[$0-1], $$[$0]);
break;
case 31:this.$ = yy.Operation($$[$0-2], $$[$0-1], $$[$0]);
break;
case 32:this.$ = yy.Operation($$[$0-2], $$[$0-1], $$[$0]);
break;
case 33:this.$ = yy.Operation($$[$0-2], $$[$0-1], $$[$0]);
break;
case 34:this.$ = $$[$0-1];
break;
case 35:this.$ = $$[$0];
break;
case 36:this.$ = $$[$0];
break;
case 37:this.$ = $$[$0];
break;
case 38:this.$ = $$[$0];
break;
case 39:this.$ = yy.Javascript($$[$0].slice(3, -3));
break;
case 40:this.$ = yy.Variable($$[$0], null);
break;
case 41:this.$ = yy.Variable($$[$0-2], $$[$0]);
break;
case 42:this.$ = yy.Variable($$[$0-2], $$[$0]);
break;
case 43:this.$ = $$[$0];
break;
case 44:this.$ = yy.UnitList([$$[$0-1], $$[$0]]);
break;
case 45:this.$ = $$[$0]; $$[$0].units.unshift($$[$0-1]);
break;
case 46:this.$ = yy.UnitList([$$[$0-2]]);
break;
case 47:this.$ = yy.ArrowApplication($$[$0-2], $$[$0]);
break;
case 48:this.$ = yy.ArrowApplication($$[$0-2], yy.UnitList([$$[$0]]));
break;
case 49:this.$ = new yy.TableAccess($$[$0-2], $$[$0]);
break;
case 50:this.$ = new yy.TableAccess($$[$0-2], $$[$0]);
break;
case 51:this.$ = $$[$0];
break;
case 52:this.$ = $$[$0];
break;
case 53:this.$ = yy.Regex($$[$0]);
break;
case 54:this.$ = Number($$[$0]);
break;
case 55:this.$ = Number($$[$0]);
break;
case 56:this.$ = yy.String($$[$0]);
break;
case 57:this.$ = new yy.Table([], false);
break;
case 58:this.$ = new yy.Table([], true);
break;
case 59:this.$ = new yy.Table([], true);
break;
case 60:this.$ = new yy.Table($$[$0-1], false);
break;
case 61:this.$ = $$[$0];
break;
case 62:this.$ = $$[$0-1];
break;
case 63:this.$ = $$[$0];
break;
case 64:this.$ = $$[$0-1];
break;
case 65:this.$ = [$$[$0]];
break;
case 66:this.$ = $$[$0-2]; $$[$0-2].push($$[$0]);
break;
case 67:this.$ = $$[$0-2]; $$[$0-2].push($$[$0]);
break;
case 68:this.$ = $$[$0-3]; $$[$0-3].push($$[$0]);
break;
case 69:this.$ = yy.Field(null, $$[$0]);
break;
case 70:this.$ = yy.Field($$[$0-2], $$[$0]);
break;
case 71:this.$ = yy.Field($$[$0-2], $$[$0]);
break;
case 72:this.$ = yy.Field($$[$0-2], $$[$0]);
break;
case 73:this.$ = yy.Field($$[$0-2], $$[$0]);
break;
case 74:this.$ = yy.RetLambda([yy.Variable('_it', null)], $$[$0-1]);
break;
case 75:this.$ = yy.RetLambda([], $$[$0-1]);
break;
case 76:this.$ = yy.RetLambda($$[$0-3].units, $$[$0-1]);
break;
case 77:this.$ = yy.RetLambda([$$[$0-3]], $$[$0-1]);
break;
case 78:this.$ = yy.Lambda([yy.Variable('_it', null)], $$[$0-2]);
break;
case 79:this.$ = yy.Lambda([], $$[$0-2]);
break;
case 80:this.$ = yy.Lambda($$[$0-5].units, $$[$0-2]);
break;
case 81:this.$ = yy.Lambda([$$[$0-5]], $$[$0-2]);
break;
case 82:this.$ = $$[$0];
break;
case 83:this.$ = yy.Operation($$[$0-2], $$[$0-1], $$[$0]);
break;
case 84:this.$ = yy.Operation($$[$0-2], $$[$0-1], $$[$0]);
break;
case 85:this.$ = $$[$0];
break;
case 86:this.$ = yy.Operation($$[$0-2], $$[$0-1], $$[$0]);
break;
case 87:this.$ = yy.Operation($$[$0-2], $$[$0-1], $$[$0]);
break;
case 88:this.$ = yy.Operation($$[$0-2], $$[$0-1], $$[$0]);
break;
case 89:this.$ = yy.Operation(null, $$[$0-1], $$[$0]);
break;
case 90:this.$ = $$[$0];
break;
}
},
table: [{3:1,4:2,5:[1,3],6:4,7:5,8:6,9:[1,7],10:8,12:10,13:9,14:14,15:15,16:16,17:23,18:24,19:25,21:13,24:[1,17],26:18,27:19,28:20,29:[1,21],30:[1,11],33:12,37:27,38:28,39:[1,29],40:[1,31],41:[1,32],42:[1,33],43:[1,34],50:[1,26],53:22,55:[1,35],56:30},{1:[3]},{5:[1,36]},{1:[2,2]},{5:[2,3],7:37,9:[1,7]},{6:38,8:6,9:[1,39],10:8,12:10,13:9,14:14,15:15,16:16,17:23,18:24,19:25,21:13,24:[1,17],26:18,27:19,28:20,29:[1,21],30:[1,11],33:12,37:27,38:28,39:[1,29],40:[1,31],41:[1,32],42:[1,33],43:[1,34],50:[1,26],53:22,55:[1,35],56:30},{5:[2,7],9:[2,7]},{5:[2,9],9:[2,9],24:[2,9],29:[2,9],30:[2,9],39:[2,9],40:[2,9],41:[2,9],42:[2,9],43:[2,9],44:[2,9],50:[2,9],51:[2,9],55:[2,9]},{5:[2,37],9:[2,37],11:[1,40],20:[2,37],22:[2,37],23:[2,37],24:[2,37],29:[2,37],30:[2,37],34:[2,37],35:[2,37],36:[2,37],39:[2,37],40:[2,37],41:[2,37],42:[2,37],43:[2,37],50:[2,37],51:[2,37],52:[2,37],54:[2,37],55:[2,37],57:[2,37],58:[2,37],59:[2,37]},{5:[2,14],9:[2,14],11:[1,41],34:[2,14],51:[2,14]},{5:[2,13],9:[2,13],34:[1,42],51:[2,13]},{5:[2,40],9:[2,40],11:[2,40],20:[2,40],22:[2,40],23:[2,40],24:[2,40],25:[2,40],29:[2,40],30:[2,40],31:[1,43],34:[2,40],35:[2,40],36:[2,40],39:[2,40],40:[2,40],41:[2,40],42:[2,40],43:[2,40],44:[2,40],49:[2,40],50:[2,40],51:[2,40],52:[2,40],54:[2,40],55:[2,40],57:[2,40],58:[2,40],59:[2,40]},{5:[2,43],9:[2,43],11:[2,43],20:[2,43],22:[2,43],23:[2,43],24:[2,43],25:[2,43],29:[2,43],30:[2,43],34:[2,43],35:[2,43],36:[2,43],39:[2,43],40:[2,43],41:[2,43],42:[2,43],43:[2,43],44:[2,43],49:[2,43],50:[2,43],51:[2,43],52:[2,43],54:[2,43],55:[2,43],57:[2,43],58:[2,43],59:[2,43]},{5:[2,90],9:[2,90],10:51,13:45,20:[1,49],21:44,22:[1,50],23:[2,90],24:[1,48],25:[2,90],26:18,27:19,28:20,29:[1,21],30:[1,11],33:12,34:[2,90],35:[1,46],36:[1,47],37:27,38:28,39:[1,29],40:[1,31],41:[1,32],42:[1,33],43:[1,34],44:[2,90],49:[2,90],50:[1,26],51:[2,90],54:[2,90],55:[2,90],57:[2,90],58:[2,90],59:[2,90]},{5:[2,15],9:[2,15],23:[1,54],25:[2,15],34:[2,15],44:[2,15],49:[2,15],51:[2,15],54:[1,52],55:[1,53]},{5:[2,16],9:[2,16],25:[2,16],34:[2,16],44:[2,16],49:[2,16],51:[2,16]},{5:[2,17],9:[2,17],25:[2,17],34:[2,17],44:[2,17],49:[2,17],51:[2,17]},{10:51,12:55,13:56,14:14,15:15,16:16,17:23,18:24,19:25,21:13,24:[1,17],26:18,27:19,28:20,29:[1,21],30:[1,11],33:12,37:27,38:28,39:[1,29],40:[1,31],41:[1,32],42:[1,33],43:[1,34],50:[1,26],53:22,55:[1,35],56:30},{5:[2,35],9:[2,35],11:[2,35],20:[2,35],22:[2,35],23:[2,35],24:[2,35],25:[2,35],29:[2,35],30:[2,35],34:[2,35],35:[2,35],36:[2,35],39:[2,35],40:[2,35],41:[2,35],42:[2,35],43:[2,35],44:[2,35],49:[2,35],50:[2,35],51:[2,35],52:[2,35],54:[2,35],55:[2,35],57:[2,35],58:[2,35],59:[2,35]},{5:[2,36],9:[2,36],11:[2,36],20:[2,36],22:[2,36],23:[2,36],24:[2,36],25:[2,36],29:[2,36],30:[2,36],34:[2,36],35:[2,36],36:[2,36],39:[2,36],40:[2,36],41:[2,36],42:[2,36],43:[2,36],44:[2,36],49:[2,36],50:[2,36],51:[2,36],52:[2,36],54:[2,36],55:[2,36],57:[2,36],58:[2,36],59:[2,36]},{5:[2,38],9:[2,38],11:[2,38],20:[2,38],22:[2,38],23:[2,38],24:[2,38],25:[2,38],29:[2,38],30:[2,38],34:[2,38],35:[2,38],36:[2,38],39:[2,38],40:[2,38],41:[2,38],42:[2,38],43:[2,38],44:[2,38],49:[2,38],50:[2,38],51:[2,38],52:[2,38],54:[2,38],55:[2,38],57:[2,38],58:[2,38],59:[2,38]},{5:[2,39],9:[2,39],11:[2,39],20:[2,39],22:[2,39],23:[2,39],24:[2,39],25:[2,39],29:[2,39],30:[2,39],34:[2,39],35:[2,39],36:[2,39],39:[2,39],40:[2,39],41:[2,39],42:[2,39],43:[2,39],44:[2,39],49:[2,39],50:[2,39],51:[2,39],52:[2,39],54:[2,39],55:[2,39],57:[2,39],58:[2,39],59:[2,39]},{5:[2,82],9:[2,82],20:[2,82],22:[2,82],23:[2,82],25:[2,82],34:[2,82],44:[2,82],49:[2,82],51:[2,82],54:[2,82],55:[2,82],57:[1,57],58:[1,58],59:[1,59]},{5:[2,18],9:[2,18],20:[1,60],22:[1,61],25:[2,18],34:[2,18],44:[2,18],49:[2,18],51:[2,18]},{5:[2,19],9:[2,19],20:[1,62],25:[2,19],34:[2,19],44:[2,19],49:[2,19],51:[2,19]},{5:[2,20],9:[2,20],22:[1,63],25:[2,20],34:[2,20],44:[2,20],49:[2,20],51:[2,20]},{7:68,8:64,9:[1,7],10:8,12:10,13:66,14:14,15:15,16:16,17:23,18:24,19:25,21:67,24:[1,17],26:18,27:19,28:20,29:[1,21],30:[1,11],33:12,37:27,38:28,39:[1,29],40:[1,31],41:[1,32],42:[1,33],43:[1,34],50:[1,26],52:[1,65],53:22,55:[1,35],56:30},{5:[2,51],9:[2,51],11:[2,51],20:[2,51],22:[2,51],23:[2,51],24:[2,51],25:[2,51],29:[2,51],30:[2,51],34:[2,51],35:[2,51],36:[2,51],39:[2,51],40:[2,51],41:[2,51],42:[2,51],43:[2,51],44:[2,51],49:[2,51],50:[2,51],51:[2,51],52:[2,51],54:[2,51],55:[2,51],57:[2,51],58:[2,51],59:[2,51]},{5:[2,52],9:[2,52],11:[2,52],20:[2,52],22:[2,52],23:[2,52],24:[2,52],25:[2,52],29:[2,52],30:[2,52],34:[2,52],35:[2,52],36:[2,52],39:[2,52],40:[2,52],41:[2,52],42:[2,52],43:[2,52],44:[2,52],49:[2,52],50:[2,52],51:[2,52],52:[2,52],54:[2,52],55:[2,52],57:[2,52],58:[2,52],59:[2,52]},{5:[2,53],9:[2,53],11:[2,53],20:[2,53],22:[2,53],23:[2,53],24:[2,53],25:[2,53],29:[2,53],30:[2,53],34:[2,53],35:[2,53],36:[2,53],39:[2,53],40:[2,53],41:[2,53],42:[2,53],43:[2,53],44:[2,53],49:[2,53],50:[2,53],51:[2,53],52:[2,53],54:[2,53],55:[2,53],57:[2,53],58:[2,53],59:[2,53]},{5:[2,85],9:[2,85],20:[2,85],22:[2,85],23:[2,85],25:[2,85],34:[2,85],44:[2,85],49:[2,85],51:[2,85],54:[2,85],55:[2,85],57:[2,85],58:[2,85],59:[2,85]},{5:[2,54],9:[2,54],11:[2,54],20:[2,54],22:[2,54],23:[2,54],24:[2,54],25:[2,54],29:[2,54],30:[2,54],34:[2,54],35:[2,54],36:[2,54],39:[2,54],40:[2,54],41:[2,54],42:[2,54],43:[2,54],44:[2,54],45:[2,54],49:[2,54],50:[2,54],51:[2,54],52:[2,54],54:[2,54],55:[2,54],57:[2,54],58:[2,54],59:[2,54]},{5:[2,55],9:[2,55],11:[2,55],20:[2,55],22:[2,55],23:[2,55],24:[2,55],25:[2,55],29:[2,55],30:[2,55],34:[2,55],35:[2,55],36:[2,55],39:[2,55],40:[2,55],41:[2,55],42:[2,55],43:[2,55],44:[2,55],45:[2,55],49:[2,55],50:[2,55],51:[2,55],52:[2,55],54:[2,55],55:[2,55],57:[2,55],58:[2,55],59:[2,55]},{5:[2,56],9:[2,56],11:[2,56],20:[2,56],22:[2,56],23:[2,56],24:[2,56],25:[2,56],29:[2,56],30:[2,56],34:[2,56],35:[2,56],36:[2,56],39:[2,56],40:[2,56],41:[2,56],42:[2,56],43:[2,56],44:[2,56],45:[2,56],49:[2,56],50:[2,56],51:[2,56],52:[2,56],54:[2,56],55:[2,56],57:[2,56],58:[2,56],59:[2,56]},{7:74,9:[1,7],10:51,11:[1,71],12:76,13:56,14:14,15:15,16:16,17:23,18:24,19:25,21:13,24:[1,17],26:18,27:19,28:20,29:[1,21],30:[1,77],33:12,37:78,38:28,39:[1,29],40:[1,31],41:[1,32],42:[1,33],43:[1,34],44:[1,69],45:[1,70],46:72,47:73,48:75,50:[1,26],53:22,55:[1,35],56:30},{10:51,21:80,24:[1,17],26:18,27:19,28:20,29:[1,21],30:[1,11],33:12,37:27,38:28,39:[1,29],40:[1,31],41:[1,32],42:[1,33],43:[1,34],50:[1,26],55:[1,35],56:79},{1:[2,1]},{5:[2,4],8:81,9:[1,39],10:8,12:10,13:9,14:14,15:15,16:16,17:23,18:24,19:25,21:13,24:[1,17],26:18,27:19,28:20,29:[1,21],30:[1,11],33:12,37:27,38:28,39:[1,29],40:[1,31],41:[1,32],42:[1,33],43:[1,34],50:[1,26],53:22,55:[1,35],56:30},{5:[2,5],7:82,9:[1,7]},{5:[2,10],9:[2,10],24:[2,10],29:[2,10],30:[2,10],39:[2,10],40:[2,10],41:[2,10],42:[2,10],43:[2,10],44:[2,10],50:[2,10],51:[2,10],55:[2,10]},{10:51,12:83,13:56,14:14,15:15,16:16,17:23,18:24,19:25,21:13,24:[1,17],26:18,27:19,28:20,29:[1,21],30:[1,11],33:12,37:27,38:28,39:[1,29],40:[1,31],41:[1,32],42:[1,33],43:[1,34],50:[1,26],53:22,55:[1,35],56:30},{10:51,12:84,13:56,14:14,15:15,16:16,17:23,18:24,19:25,21:13,24:[1,17],26:18,27:19,28:20,29:[1,21],30:[1,11],33:12,37:27,38:28,39:[1,29],40:[1,31],41:[1,32],42:[1,33],43:[1,34],50:[1,26],53:22,55:[1,35],56:30},{10:51,13:85,21:86,24:[1,17],26:18,27:19,28:20,29:[1,21],30:[1,11],33:12,37:27,38:28,39:[1,29],40:[1,31],41:[1,32],42:[1,33],43:[1,34],50:[1,26]},{10:51,21:87,24:[1,17],26:18,27:19,28:20,29:[1,21],30:[1,11],32:[1,88],33:12,37:27,38:28,39:[1,29],40:[1,31],41:[1,32],42:[1,33],43:[1,34],50:[1,26]},{5:[2,44],9:[2,44],10:51,11:[2,44],13:45,21:44,24:[1,48],25:[2,44],26:18,27:19,28:20,29:[1,21],30:[1,11],33:12,34:[2,44],35:[1,46],36:[1,47],37:27,38:28,39:[1,29],40:[1,31],41:[1,32],42:[1,33],43:[1,34],44:[2,44],49:[2,44],50:[1,26],51:[2,44],52:[2,44]},{5:[2,45],9:[2,45],11:[2,45],25:[2,45],34:[2,45],44:[2,45],49:[2,45],51:[2,45],52:[2,45]},{30:[1,89]},{10:51,21:90,24:[1,17],26:18,27:19,28:20,29:[1,21],30:[1,11],33:12,37:27,38:28,39:[1,29],40:[1,31],41:[1,32],42:[1,33],43:[1,34],50:[1,26]},{10:51,12:55,13:56,14:14,15:15,16:16,17:23,18:24,19:25,21:13,24:[1,17],25:[1,91],26:18,27:19,28:20,29:[1,21],30:[1,11],33:12,37:27,38:28,39:[1,29],40:[1,31],41:[1,32],42:[1,33],43:[1,34],50:[1,26],53:22,55:[1,35],56:30},{10:51,14:94,17:92,21:93,24:[1,17],26:18,27:19,28:20,29:[1,21],30:[1,11],33:12,37:27,38:28,39:[1,29],40:[1,31],41:[1,32],42:[1,33],43:[1,34],50:[1,26],53:22,55:[1,35],56:30},{10:51,14:94,17:95,21:96,24:[1,17],26:18,27:19,28:20,29:[1,21],30:[1,11],33:12,37:27,38:28,39:[1,29],40:[1,31],41:[1,32],42:[1,33],43:[1,34],50:[1,26],53:22,55:[1,35],56:30},{5:[2,37],9:[2,37],11:[2,37],20:[2,37],22:[2,37],23:[2,37],24:[2,37],25:[2,37],29:[2,37],30:[2,37],34:[2,37],35:[2,37],36:[2,37],39:[2,37],40:[2,37],41:[2,37],42:[2,37],43:[2,37],44:[2,37],49:[2,37],50:[2,37],51:[2,37],52:[2,37],54:[2,37],55:[2,37],57:[2,37],58:[2,37],59:[2,37]},{10:51,14:97,21:80,24:[1,17],26:18,27:19,28:20,29:[1,21],30:[1,11],33:12,37:27,38:28,39:[1,29],40:[1,31],41:[1,32],42:[1,33],43:[1,34],50:[1,26],53:22,55:[1,35],56:30},{10:51,14:98,21:80,24:[1,17],26:18,27:19,28:20,29:[1,21],30:[1,11],33:12,37:27,38:28,39:[1,29],40:[1,31],41:[1,32],42:[1,33],43:[1,34],50:[1,26],53:22,55:[1,35],56:30},{10:51,14:99,21:80,24:[1,17],26:18,27:19,28:20,29:[1,21],30:[1,11],33:12,37:27,38:28,39:[1,29],40:[1,31],41:[1,32],42:[1,33],43:[1,34],50:[1,26],53:22,55:[1,35],56:30},{25:[1,100],34:[1,42]},{5:[2,14],9:[2,14],25:[2,14],34:[2,14],44:[2,14],49:[2,14],51:[2,14]},{10:51,21:80,24:[1,17],26:18,27:19,28:20,29:[1,21],30:[1,11],33:12,37:27,38:28,39:[1,29],40:[1,31],41:[1,32],42:[1,33],43:[1,34],50:[1,26],53:101,55:[1,35],56:30},{10:51,21:80,24:[1,17],26:18,27:19,28:20,29:[1,21],30:[1,11],33:12,37:27,38:28,39:[1,29],40:[1,31],41:[1,32],42:[1,33],43:[1,34],50:[1,26],53:102,55:[1,35],56:30},{10:51,21:80,24:[1,17],26:18,27:19,28:20,29:[1,21],30:[1,11],33:12,37:27,38:28,39:[1,29],40:[1,31],41:[1,32],42:[1,33],43:[1,34],50:[1,26],53:103,55:[1,35],56:30},{10:51,14:94,17:104,21:105,24:[1,17],26:18,27:19,28:20,29:[1,21],30:[1,11],33:12,37:27,38:28,39:[1,29],40:[1,31],41:[1,32],42:[1,33],43:[1,34],50:[1,26],53:22,55:[1,35],56:30},{10:51,14:94,17:106,21:107,24:[1,17],26:18,27:19,28:20,29:[1,21],30:[1,11],33:12,37:27,38:28,39:[1,29],40:[1,31],41:[1,32],42:[1,33],43:[1,34],50:[1,26],53:22,55:[1,35],56:30},{10:51,14:94,17:108,21:109,24:[1,17],26:18,27:19,28:20,29:[1,21],30:[1,11],33:12,37:27,38:28,39:[1,29],40:[1,31],41:[1,32],42:[1,33],43:[1,34],50:[1,26],53:22,55:[1,35],56:30},{10:51,14:94,17:110,21:111,24:[1,17],26:18,27:19,28:20,29:[1,21],30:[1,11],33:12,37:27,38:28,39:[1,29],40:[1,31],41:[1,32],42:[1,33],43:[1,34],50:[1,26],53:22,55:[1,35],56:30},{51:[1,112]},{7:114,8:113,9:[1,7],10:8,12:10,13:9,14:14,15:15,16:16,17:23,18:24,19:25,21:13,24:[1,17],26:18,27:19,28:20,29:[1,21],30:[1,11],33:12,37:27,38:28,39:[1,29],40:[1,31],41:[1,32],42:[1,33],43:[1,34],50:[1,26],53:22,55:[1,35],56:30},{11:[1,41],34:[2,14],51:[2,14],52:[1,115]},{10:51,13:45,20:[1,49],21:44,22:[1,50],23:[2,90],24:[1,48],26:18,27:19,28:20,29:[1,21],30:[1,11],33:12,34:[2,90],35:[1,46],36:[1,47],37:27,38:28,39:[1,29],40:[1,31],41:[1,32],42:[1,33],43:[1,34],50:[1,26],51:[2,90],52:[1,116],54:[2,90],55:[2,90],57:[2,90],58:[2,90],59:[2,90]},{6:117,8:6,9:[1,39],10:8,12:10,13:9,14:14,15:15,16:16,17:23,18:24,19:25,21:13,24:[1,17],26:18,27:19,28:20,29:[1,21],30:[1,11],33:12,37:27,38:28,39:[1,29],40:[1,31],41:[1,32],42:[1,33],43:[1,34],50:[1,26],53:22,55:[1,35],56:30},{5:[2,57],9:[2,57],11:[2,57],20:[2,57],22:[2,57],23:[2,57],24:[2,57],25:[2,57],29:[2,57],30:[2,57],34:[2,57],35:[2,57],36:[2,57],39:[2,57],40:[2,57],41:[2,57],42:[2,57],43:[2,57],44:[2,57],49:[2,57],50:[2,57],51:[2,57],52:[2,57],54:[2,57],55:[2,57],57:[2,57],58:[2,57],59:[2,57]},{44:[1,118]},{44:[1,119]},{44:[1,120]},{7:121,9:[1,7],44:[2,61],49:[1,122]},{9:[1,39],10:51,12:76,13:56,14:14,15:15,16:16,17:23,18:24,19:25,21:13,24:[1,17],26:18,27:19,28:20,29:[1,21],30:[1,77],33:12,37:78,38:28,39:[1,29],40:[1,31],41:[1,32],42:[1,33],43:[1,34],47:123,48:75,50:[1,26],53:22,55:[1,35],56:30},{9:[2,65],44:[2,65],49:[2,65]},{9:[2,69],34:[1,42],44:[2,69],49:[2,69]},{9:[2,40],11:[1,125],20:[2,40],22:[2,40],23:[2,40],24:[2,40],29:[2,40],30:[2,40],31:[1,43],34:[2,40],35:[2,40],36:[2,40],39:[2,40],40:[2,40],41:[2,40],42:[2,40],43:[2,40],44:[2,40],45:[1,124],49:[2,40],50:[2,40],54:[2,40],55:[2,40],57:[2,40],58:[2,40],59:[2,40]},{9:[2,51],11:[1,127],20:[2,51],22:[2,51],23:[2,51],24:[2,51],29:[2,51],30:[2,51],34:[2,51],35:[2,51],36:[2,51],39:[2,51],40:[2,51],41:[2,51],42:[2,51],43:[2,51],44:[2,51],45:[1,126],49:[2,51],50:[2,51],54:[2,51],55:[2,51],57:[2,51],58:[2,51],59:[2,51]},{5:[2,89],9:[2,89],20:[2,89],22:[2,89],23:[2,89],25:[2,89],34:[2,89],44:[2,89],49:[2,89],51:[2,89],54:[2,89],55:[2,89],57:[2,89],58:[2,89],59:[2,89]},{5:[2,90],9:[2,90],20:[2,90],22:[2,90],23:[2,90],24:[1,128],25:[2,90],34:[2,90],35:[1,46],36:[1,47],44:[2,90],49:[2,90],51:[2,90],54:[2,90],55:[2,90],57:[2,90],58:[2,90],59:[2,90]},{5:[2,8],9:[2,8]},{5:[2,6],8:81,9:[1,39],10:8,12:10,13:9,14:14,15:15,16:16,17:23,18:24,19:25,21:13,24:[1,17],26:18,27:19,28:20,29:[1,21],30:[1,11],33:12,37:27,38:28,39:[1,29],40:[1,31],41:[1,32],42:[1,33],43:[1,34],50:[1,26],53:22,55:[1,35],56:30},{5:[2,11],9:[2,11],34:[1,42],51:[2,11]},{5:[2,12],9:[2,12],34:[1,42],51:[2,12]},{5:[2,47],9:[2,47],25:[2,47],34:[2,47],44:[2,47],49:[2,47],51:[2,47]},{5:[2,48],9:[2,48],10:51,13:45,21:44,24:[1,48],25:[2,48],26:18,27:19,28:20,29:[1,21],30:[1,11],33:12,34:[2,48],35:[1,46],36:[1,47],37:27,38:28,39:[1,29],40:[1,31],41:[1,32],42:[1,33],43:[1,34],44:[2,48],49:[2,48],50:[1,26],51:[2,48]},{5:[2,41],9:[2,41],11:[2,41],20:[2,41],22:[2,41],23:[2,41],24:[2,41],25:[2,41],29:[2,41],30:[2,41],34:[2,41],35:[2,41],36:[2,41],39:[2,41],40:[2,41],41:[2,41],42:[2,41],43:[2,41],44:[2,41],49:[2,41],50:[2,41],51:[2,41],52:[2,41],54:[2,41],55:[2,41],57:[2,41],58:[2,41],59:[2,41]},{5:[2,42],9:[2,42],11:[2,42],20:[2,42],22:[2,42],23:[2,42],24:[2,42],25:[2,42],29:[2,42],30:[2,42],34:[2,42],35:[2,42],36:[2,42],39:[2,42],40:[2,42],41:[2,42],42:[2,42],43:[2,42],44:[2,42],49:[2,42],50:[2,42],51:[2,42],52:[2,42],54:[2,42],55:[2,42],57:[2,42],58:[2,42],59:[2,42]},{5:[2,49],9:[2,49],11:[2,49],20:[2,49],22:[2,49],23:[2,49],24:[2,49],25:[2,49],29:[2,49],30:[2,49],34:[2,49],35:[2,49],36:[2,49],39:[2,49],40:[2,49],41:[2,49],42:[2,49],43:[2,49],44:[2,49],49:[2,49],50:[2,49],51:[2,49],52:[2,49],54:[2,49],55:[2,49],57:[2,49],58:[2,49],59:[2,49]},{5:[2,50],9:[2,50],11:[2,50],20:[2,50],22:[2,50],23:[2,50],24:[2,50],25:[2,50],29:[2,50],30:[2,50],34:[2,50],35:[2,50],36:[2,50],39:[2,50],40:[2,50],41:[2,50],42:[2,50],43:[2,50],44:[2,50],49:[2,50],50:[2,50],51:[2,50],52:[2,50],54:[2,50],55:[2,50],57:[2,50],58:[2,50],59:[2,50]},{5:[2,46],9:[2,46],11:[2,46],20:[2,46],22:[2,46],23:[2,46],24:[2,46],25:[2,46],29:[2,46],30:[2,46],34:[2,46],35:[2,46],36:[2,46],39:[2,46],40:[2,46],41:[2,46],42:[2,46],43:[2,46],44:[2,46],49:[2,46],50:[2,46],51:[2,46],52:[2,46],54:[2,46],55:[2,46],57:[2,46],58:[2,46],59:[2,46]},{5:[2,23],9:[2,23],20:[2,23],25:[2,23],34:[2,23],44:[2,23],49:[2,23],51:[2,23]},{5:[2,24],9:[2,24],20:[2,24],23:[2,90],24:[1,128],25:[2,24],34:[2,24],35:[1,46],36:[1,47],44:[2,24],49:[2,24],51:[2,24],54:[2,90],55:[2,90],57:[2,90],58:[2,90],59:[2,90]},{23:[1,54],54:[1,52],55:[1,53]},{5:[2,29],9:[2,29],22:[2,29],25:[2,29],34:[2,29],44:[2,29],49:[2,29],51:[2,29]},{5:[2,30],9:[2,30],22:[2,30],23:[2,90],24:[1,128],25:[2,30],34:[2,30],35:[1,46],36:[1,47],44:[2,30],49:[2,30],51:[2,30],54:[2,90],55:[2,90],57:[2,90],58:[2,90],59:[2,90]},{5:[2,83],9:[2,83],20:[2,83],22:[2,83],23:[2,83],25:[2,83],34:[2,83],44:[2,83],49:[2,83],51:[2,83],54:[2,83],55:[2,83]},{5:[2,84],9:[2,84],20:[2,84],22:[2,84],23:[2,84],25:[2,84],34:[2,84],44:[2,84],49:[2,84],51:[2,84],54:[2,84],55:[2,84]},{5:[2,33],9:[2,33],20:[2,33],22:[2,33],25:[2,33],34:[2,33],44:[2,33],49:[2,33],51:[2,33],54:[1,52],55:[1,53]},{5:[2,34],9:[2,34],11:[2,34],20:[2,34],22:[2,34],23:[2,34],24:[2,34],25:[2,34],29:[2,34],30:[2,34],34:[2,34],35:[2,34],36:[2,34],39:[2,34],40:[2,34],41:[2,34],42:[2,34],43:[2,34],44:[2,34],49:[2,34],50:[2,34],51:[2,34],52:[2,34],54:[2,34],55:[2,34],57:[2,34],58:[2,34],59:[2,34]},{5:[2,86],9:[2,86],20:[2,86],22:[2,86],23:[2,86],25:[2,86],34:[2,86],44:[2,86],49:[2,86],51:[2,86],54:[2,86],55:[2,86],57:[2,86],58:[2,86],59:[2,86]},{5:[2,87],9:[2,87],20:[2,87],22:[2,87],23:[2,87],25:[2,87],34:[2,87],44:[2,87],49:[2,87],51:[2,87],54:[2,87],55:[2,87],57:[2,87],58:[2,87],59:[2,87]},{5:[2,88],9:[2,88],20:[2,88],22:[2,88],23:[2,88],25:[2,88],34:[2,88],44:[2,88],49:[2,88],51:[2,88],54:[2,88],55:[2,88],57:[2,88],58:[2,88],59:[2,88]},{5:[2,21],9:[2,21],20:[2,21],25:[2,21],34:[2,21],44:[2,21],49:[2,21],51:[2,21]},{5:[2,22],9:[2,22],20:[2,22],23:[2,90],24:[1,128],25:[2,22],34:[2,22],35:[1,46],36:[1,47],44:[2,22],49:[2,22],51:[2,22],54:[2,90],55:[2,90],57:[2,90],58:[2,90],59:[2,90]},{5:[2,27],9:[2,27],22:[2,27],25:[2,27],34:[2,27],44:[2,27],49:[2,27],51:[2,27]},{5:[2,28],9:[2,28],22:[2,28],23:[2,90],24:[1,128],25:[2,28],34:[2,28],35:[1,46],36:[1,47],44:[2,28],49:[2,28],51:[2,28],54:[2,90],55:[2,90],57:[2,90],58:[2,90],59:[2,90]},{5:[2,25],9:[2,25],20:[2,25],25:[2,25],34:[2,25],44:[2,25],49:[2,25],51:[2,25]},{5:[2,26],9:[2,26],20:[2,26],23:[2,90],24:[1,128],25:[2,26],34:[2,26],35:[1,46],36:[1,47],44:[2,26],49:[2,26],51:[2,26],54:[2,90],55:[2,90],57:[2,90],58:[2,90],59:[2,90]},{5:[2,31],9:[2,31],22:[2,31],25:[2,31],34:[2,31],44:[2,31],49:[2,31],51:[2,31]},{5:[2,32],9:[2,32],22:[2,32],23:[2,90],24:[1,128],25:[2,32],34:[2,32],35:[1,46],36:[1,47],44:[2,32],49:[2,32],51:[2,32],54:[2,90],55:[2,90],57:[2,90],58:[2,90],59:[2,90]},{5:[2,74],9:[2,74],11:[2,74],20:[2,74],22:[2,74],23:[2,74],24:[2,74],25:[2,74],29:[2,74],30:[2,74],34:[2,74],35:[2,74],36:[2,74],39:[2,74],40:[2,74],41:[2,74],42:[2,74],43:[2,74],44:[2,74],49:[2,74],50:[2,74],51:[2,74],52:[2,74],54:[2,74],55:[2,74],57:[2,74],58:[2,74],59:[2,74]},{51:[1,129]},{6:130,8:6,9:[1,39],10:8,12:10,13:9,14:14,15:15,16:16,17:23,18:24,19:25,21:13,24:[1,17],26:18,27:19,28:20,29:[1,21],30:[1,11],33:12,37:27,38:28,39:[1,29],40:[1,31],41:[1,32],42:[1,33],43:[1,34],50:[1,26],53:22,55:[1,35],56:30},{7:132,8:131,9:[1,7],10:8,12:10,13:9,14:14,15:15,16:16,17:23,18:24,19:25,21:13,24:[1,17],26:18,27:19,28:20,29:[1,21],30:[1,11],33:12,37:27,38:28,39:[1,29],40:[1,31],41:[1,32],42:[1,33],43:[1,34],50:[1,26],53:22,55:[1,35],56:30},{7:134,8:133,9:[1,7],10:8,12:10,13:9,14:14,15:15,16:16,17:23,18:24,19:25,21:13,24:[1,17],26:18,27:19,28:20,29:[1,21],30:[1,11],33:12,37:27,38:28,39:[1,29],40:[1,31],41:[1,32],42:[1,33],43:[1,34],50:[1,26],53:22,55:[1,35],56:30},{7:135,9:[1,7]},{5:[2,58],9:[2,58],11:[2,58],20:[2,58],22:[2,58],23:[2,58],24:[2,58],25:[2,58],29:[2,58],30:[2,58],34:[2,58],35:[2,58],36:[2,58],39:[2,58],40:[2,58],41:[2,58],42:[2,58],43:[2,58],44:[2,58],49:[2,58],50:[2,58],51:[2,58],52:[2,58],54:[2,58],55:[2,58],57:[2,58],58:[2,58],59:[2,58]},{5:[2,59],9:[2,59],11:[2,59],20:[2,59],22:[2,59],23:[2,59],24:[2,59],25:[2,59],29:[2,59],30:[2,59],34:[2,59],35:[2,59],36:[2,59],39:[2,59],40:[2,59],41:[2,59],42:[2,59],43:[2,59],44:[2,59],49:[2,59],50:[2,59],51:[2,59],52:[2,59],54:[2,59],55:[2,59],57:[2,59],58:[2,59],59:[2,59]},{5:[2,60],9:[2,60],11:[2,60],20:[2,60],22:[2,60],23:[2,60],24:[2,60],25:[2,60],29:[2,60],30:[2,60],34:[2,60],35:[2,60],36:[2,60],39:[2,60],40:[2,60],41:[2,60],42:[2,60],43:[2,60],44:[2,60],49:[2,60],50:[2,60],51:[2,60],52:[2,60],54:[2,60],55:[2,60],57:[2,60],58:[2,60],59:[2,60]},{9:[1,39],10:51,12:76,13:56,14:14,15:15,16:16,17:23,18:24,19:25,21:13,24:[1,17],26:18,27:19,28:20,29:[1,21],30:[1,77],33:12,37:78,38:28,39:[1,29],40:[1,31],41:[1,32],42:[1,33],43:[1,34],44:[2,62],48:136,50:[1,26],53:22,55:[1,35],56:30},{7:138,9:[1,7],10:51,12:76,13:56,14:14,15:15,16:16,17:23,18:24,19:25,21:13,24:[1,17],26:18,27:19,28:20,29:[1,21],30:[1,77],33:12,37:78,38:28,39:[1,29],40:[1,31],41:[1,32],42:[1,33],43:[1,34],48:137,50:[1,26],53:22,55:[1,35],56:30},{7:139,9:[1,7],44:[2,63],49:[1,122]},{10:51,12:140,13:56,14:14,15:15,16:16,17:23,18:24,19:25,21:13,24:[1,17],26:18,27:19,28:20,29:[1,21],30:[1,11],33:12,37:27,38:28,39:[1,29],40:[1,31],41:[1,32],42:[1,33],43:[1,34],50:[1,26],53:22,55:[1,35],56:30},{10:51,12:141,13:56,14:14,15:15,16:16,17:23,18:24,19:25,21:13,24:[1,17],26:18,27:19,28:20,29:[1,21],30:[1,11],33:12,37:27,38:28,39:[1,29],40:[1,31],41:[1,32],42:[1,33],43:[1,34],50:[1,26],53:22,55:[1,35],56:30},{10:51,12:142,13:56,14:14,15:15,16:16,17:23,18:24,19:25,21:13,24:[1,17],26:18,27:19,28:20,29:[1,21],30:[1,11],33:12,37:27,38:28,39:[1,29],40:[1,31],41:[1,32],42:[1,33],43:[1,34],50:[1,26],53:22,55:[1,35],56:30},{10:51,12:143,13:56,14:14,15:15,16:16,17:23,18:24,19:25,21:13,24:[1,17],26:18,27:19,28:20,29:[1,21],30:[1,11],33:12,37:27,38:28,39:[1,29],40:[1,31],41:[1,32],42:[1,33],43:[1,34],50:[1,26],53:22,55:[1,35],56:30},{25:[1,91]},{5:[2,75],9:[2,75],11:[2,75],20:[2,75],22:[2,75],23:[2,75],24:[2,75],25:[2,75],29:[2,75],30:[2,75],34:[2,75],35:[2,75],36:[2,75],39:[2,75],40:[2,75],41:[2,75],42:[2,75],43:[2,75],44:[2,75],49:[2,75],50:[2,75],51:[2,75],52:[2,75],54:[2,75],55:[2,75],57:[2,75],58:[2,75],59:[2,75]},{7:144,9:[1,7]},{51:[1,145]},{6:146,8:6,9:[1,39],10:8,12:10,13:9,14:14,15:15,16:16,17:23,18:24,19:25,21:13,24:[1,17],26:18,27:19,28:20,29:[1,21],30:[1,11],33:12,37:27,38:28,39:[1,29],40:[1,31],41:[1,32],42:[1,33],43:[1,34],50:[1,26],53:22,55:[1,35],56:30},{51:[1,147]},{6:148,8:6,9:[1,39],10:8,12:10,13:9,14:14,15:15,16:16,17:23,18:24,19:25,21:13,24:[1,17],26:18,27:19,28:20,29:[1,21],30:[1,11],33:12,37:27,38:28,39:[1,29],40:[1,31],41:[1,32],42:[1,33],43:[1,34],50:[1,26],53:22,55:[1,35],56:30},{8:81,9:[1,39],10:8,12:10,13:9,14:14,15:15,16:16,17:23,18:24,19:25,21:13,24:[1,17],26:18,27:19,28:20,29:[1,21],30:[1,11],33:12,37:27,38:28,39:[1,29],40:[1,31],41:[1,32],42:[1,33],43:[1,34],50:[1,26],51:[1,149],53:22,55:[1,35],56:30},{9:[2,67],44:[2,67],49:[2,67]},{9:[2,66],44:[2,66],49:[2,66]},{9:[1,39],10:51,12:76,13:56,14:14,15:15,16:16,17:23,18:24,19:25,21:13,24:[1,17],26:18,27:19,28:20,29:[1,21],30:[1,77],33:12,37:78,38:28,39:[1,29],40:[1,31],41:[1,32],42:[1,33],43:[1,34],48:150,50:[1,26],53:22,55:[1,35],56:30},{9:[1,39],10:51,12:76,13:56,14:14,15:15,16:16,17:23,18:24,19:25,21:13,24:[1,17],26:18,27:19,28:20,29:[1,21],30:[1,77],33:12,37:78,38:28,39:[1,29],40:[1,31],41:[1,32],42:[1,33],43:[1,34],44:[2,64],48:136,50:[1,26],53:22,55:[1,35],56:30},{9:[2,70],34:[1,42],44:[2,70],49:[2,70]},{9:[2,71],34:[1,42],44:[2,71],49:[2,71]},{9:[2,72],34:[1,42],44:[2,72],49:[2,72]},{9:[2,73],34:[1,42],44:[2,73],49:[2,73]},{8:81,9:[1,39],10:8,12:10,13:9,14:14,15:15,16:16,17:23,18:24,19:25,21:13,24:[1,17],26:18,27:19,28:20,29:[1,21],30:[1,11],33:12,37:27,38:28,39:[1,29],40:[1,31],41:[1,32],42:[1,33],43:[1,34],50:[1,26],51:[1,151],53:22,55:[1,35],56:30},{5:[2,76],9:[2,76],11:[2,76],20:[2,76],22:[2,76],23:[2,76],24:[2,76],25:[2,76],29:[2,76],30:[2,76],34:[2,76],35:[2,76],36:[2,76],39:[2,76],40:[2,76],41:[2,76],42:[2,76],43:[2,76],44:[2,76],49:[2,76],50:[2,76],51:[2,76],52:[2,76],54:[2,76],55:[2,76],57:[2,76],58:[2,76],59:[2,76]},{7:152,9:[1,7]},{5:[2,77],9:[2,77],11:[2,77],20:[2,77],22:[2,77],23:[2,77],24:[2,77],25:[2,77],29:[2,77],30:[2,77],34:[2,77],35:[2,77],36:[2,77],39:[2,77],40:[2,77],41:[2,77],42:[2,77],43:[2,77],44:[2,77],49:[2,77],50:[2,77],51:[2,77],52:[2,77],54:[2,77],55:[2,77],57:[2,77],58:[2,77],59:[2,77]},{7:153,9:[1,7]},{5:[2,78],9:[2,78],11:[2,78],20:[2,78],22:[2,78],23:[2,78],24:[2,78],25:[2,78],29:[2,78],30:[2,78],34:[2,78],35:[2,78],36:[2,78],39:[2,78],40:[2,78],41:[2,78],42:[2,78],43:[2,78],44:[2,78],49:[2,78],50:[2,78],51:[2,78],52:[2,78],54:[2,78],55:[2,78],57:[2,78],58:[2,78],59:[2,78]},{9:[2,68],44:[2,68],49:[2,68]},{5:[2,79],9:[2,79],11:[2,79],20:[2,79],22:[2,79],23:[2,79],24:[2,79],25:[2,79],29:[2,79],30:[2,79],34:[2,79],35:[2,79],36:[2,79],39:[2,79],40:[2,79],41:[2,79],42:[2,79],43:[2,79],44:[2,79],49:[2,79],50:[2,79],51:[2,79],52:[2,79],54:[2,79],55:[2,79],57:[2,79],58:[2,79],59:[2,79]},{8:81,9:[1,39],10:8,12:10,13:9,14:14,15:15,16:16,17:23,18:24,19:25,21:13,24:[1,17],26:18,27:19,28:20,29:[1,21],30:[1,11],33:12,37:27,38:28,39:[1,29],40:[1,31],41:[1,32],42:[1,33],43:[1,34],50:[1,26],51:[1,154],53:22,55:[1,35],56:30},{8:81,9:[1,39],10:8,12:10,13:9,14:14,15:15,16:16,17:23,18:24,19:25,21:13,24:[1,17],26:18,27:19,28:20,29:[1,21],30:[1,11],33:12,37:27,38:28,39:[1,29],40:[1,31],41:[1,32],42:[1,33],43:[1,34],50:[1,26],51:[1,155],53:22,55:[1,35],56:30},{5:[2,80],9:[2,80],11:[2,80],20:[2,80],22:[2,80],23:[2,80],24:[2,80],25:[2,80],29:[2,80],30:[2,80],34:[2,80],35:[2,80],36:[2,80],39:[2,80],40:[2,80],41:[2,80],42:[2,80],43:[2,80],44:[2,80],49:[2,80],50:[2,80],51:[2,80],52:[2,80],54:[2,80],55:[2,80],57:[2,80],58:[2,80],59:[2,80]},{5:[2,81],9:[2,81],11:[2,81],20:[2,81],22:[2,81],23:[2,81],24:[2,81],25:[2,81],29:[2,81],30:[2,81],34:[2,81],35:[2,81],36:[2,81],39:[2,81],40:[2,81],41:[2,81],42:[2,81],43:[2,81],44:[2,81],49:[2,81],50:[2,81],51:[2,81],52:[2,81],54:[2,81],55:[2,81],57:[2,81],58:[2,81],59:[2,81]}],
defaultActions: {3:[2,2],36:[2,1]},
parseError: function parseError(str, hash) {
    if (hash.recoverable) {
        this.trace(str);
    } else {
        throw new Error(str);
    }
},
parse: function parse(input) {
    var self = this, stack = [0], vstack = [null], lstack = [], table = this.table, yytext = '', yylineno = 0, yyleng = 0, recovering = 0, TERROR = 2, EOF = 1;
    var args = lstack.slice.call(arguments, 1);
    this.lexer.setInput(input);
    this.lexer.yy = this.yy;
    this.yy.lexer = this.lexer;
    this.yy.parser = this;
    if (typeof this.lexer.yylloc == 'undefined') {
        this.lexer.yylloc = {};
    }
    var yyloc = this.lexer.yylloc;
    lstack.push(yyloc);
    var ranges = this.lexer.options && this.lexer.options.ranges;
    if (typeof this.yy.parseError === 'function') {
        this.parseError = this.yy.parseError;
    } else {
        this.parseError = Object.getPrototypeOf(this).parseError;
    }
    function popStack(n) {
        stack.length = stack.length - 2 * n;
        vstack.length = vstack.length - n;
        lstack.length = lstack.length - n;
    }
    function lex() {
        var token;
        token = self.lexer.lex() || EOF;
        if (typeof token !== 'number') {
            token = self.symbols_[token] || token;
        }
        return token;
    }
    var symbol, preErrorSymbol, state, action, a, r, yyval = {}, p, len, newState, expected;
    while (true) {
        state = stack[stack.length - 1];
        if (this.defaultActions[state]) {
            action = this.defaultActions[state];
        } else {
            if (symbol === null || typeof symbol == 'undefined') {
                symbol = lex();
            }
            action = table[state] && table[state][symbol];
        }
                    if (typeof action === 'undefined' || !action.length || !action[0]) {
                var errStr = '';
                expected = [];
                for (p in table[state]) {
                    if (this.terminals_[p] && p > TERROR) {
                        expected.push('\'' + this.terminals_[p] + '\'');
                    }
                }
                if (this.lexer.showPosition) {
                    errStr = 'Parse error on line ' + (yylineno + 1) + ':\n' + this.lexer.showPosition() + '\nExpecting ' + expected.join(', ') + ', got \'' + (this.terminals_[symbol] || symbol) + '\'';
                } else {
                    errStr = 'Parse error on line ' + (yylineno + 1) + ': Unexpected ' + (symbol == EOF ? 'end of input' : '\'' + (this.terminals_[symbol] || symbol) + '\'');
                }
                this.parseError(errStr, {
                    text: this.lexer.match,
                    token: this.terminals_[symbol] || symbol,
                    line: this.lexer.yylineno,
                    loc: yyloc,
                    expected: expected
                });
            }
        if (action[0] instanceof Array && action.length > 1) {
            throw new Error('Parse Error: multiple actions possible at state: ' + state + ', token: ' + symbol);
        }
        switch (action[0]) {
        case 1:
            stack.push(symbol);
            vstack.push(this.lexer.yytext);
            lstack.push(this.lexer.yylloc);
            stack.push(action[1]);
            symbol = null;
            if (!preErrorSymbol) {
                yyleng = this.lexer.yyleng;
                yytext = this.lexer.yytext;
                yylineno = this.lexer.yylineno;
                yyloc = this.lexer.yylloc;
                if (recovering > 0) {
                    recovering--;
                }
            } else {
                symbol = preErrorSymbol;
                preErrorSymbol = null;
            }
            break;
        case 2:
            len = this.productions_[action[1]][1];
            yyval.$ = vstack[vstack.length - len];
            yyval._$ = {
                first_line: lstack[lstack.length - (len || 1)].first_line,
                last_line: lstack[lstack.length - 1].last_line,
                first_column: lstack[lstack.length - (len || 1)].first_column,
                last_column: lstack[lstack.length - 1].last_column
            };
            if (ranges) {
                yyval._$.range = [
                    lstack[lstack.length - (len || 1)].range[0],
                    lstack[lstack.length - 1].range[1]
                ];
            }
            r = this.performAction.apply(yyval, [
                yytext,
                yyleng,
                yylineno,
                this.yy,
                action[1],
                vstack,
                lstack
            ].concat(args));
            if (typeof r !== 'undefined') {
                return r;
            }
            if (len) {
                stack = stack.slice(0, -1 * len * 2);
                vstack = vstack.slice(0, -1 * len);
                lstack = lstack.slice(0, -1 * len);
            }
            stack.push(this.productions_[action[1]][0]);
            vstack.push(yyval.$);
            lstack.push(yyval._$);
            newState = table[stack[stack.length - 2]][stack[stack.length - 1]];
            stack.push(newState);
            break;
        case 3:
            return true;
        }
    }
    return true;
}};
/* generated by jison-lex 0.2.1 */
var lexer = (function(){
var lexer = {

EOF:1,

parseError:function parseError(str, hash) {
        if (this.yy.parser) {
            this.yy.parser.parseError(str, hash);
        } else {
            throw new Error(str);
        }
    },

// resets the lexer, sets new input
setInput:function (input) {
        this._input = input;
        this._more = this._backtrack = this.done = false;
        this.yylineno = this.yyleng = 0;
        this.yytext = this.matched = this.match = '';
        this.conditionStack = ['INITIAL'];
        this.yylloc = {
            first_line: 1,
            first_column: 0,
            last_line: 1,
            last_column: 0
        };
        if (this.options.ranges) {
            this.yylloc.range = [0,0];
        }
        this.offset = 0;
        return this;
    },

// consumes and returns one char from the input
input:function () {
        var ch = this._input[0];
        this.yytext += ch;
        this.yyleng++;
        this.offset++;
        this.match += ch;
        this.matched += ch;
        var lines = ch.match(/(?:\r\n?|\n).*/g);
        if (lines) {
            this.yylineno++;
            this.yylloc.last_line++;
        } else {
            this.yylloc.last_column++;
        }
        if (this.options.ranges) {
            this.yylloc.range[1]++;
        }

        this._input = this._input.slice(1);
        return ch;
    },

// unshifts one char (or a string) into the input
unput:function (ch) {
        var len = ch.length;
        var lines = ch.split(/(?:\r\n?|\n)/g);

        this._input = ch + this._input;
        this.yytext = this.yytext.substr(0, this.yytext.length - len - 1);
        //this.yyleng -= len;
        this.offset -= len;
        var oldLines = this.match.split(/(?:\r\n?|\n)/g);
        this.match = this.match.substr(0, this.match.length - 1);
        this.matched = this.matched.substr(0, this.matched.length - 1);

        if (lines.length - 1) {
            this.yylineno -= lines.length - 1;
        }
        var r = this.yylloc.range;

        this.yylloc = {
            first_line: this.yylloc.first_line,
            last_line: this.yylineno + 1,
            first_column: this.yylloc.first_column,
            last_column: lines ?
                (lines.length === oldLines.length ? this.yylloc.first_column : 0)
                 + oldLines[oldLines.length - lines.length].length - lines[0].length :
              this.yylloc.first_column - len
        };

        if (this.options.ranges) {
            this.yylloc.range = [r[0], r[0] + this.yyleng - len];
        }
        this.yyleng = this.yytext.length;
        return this;
    },

// When called from action, caches matched text and appends it on next action
more:function () {
        this._more = true;
        return this;
    },

// When called from action, signals the lexer that this rule fails to match the input, so the next matching rule (regex) should be tested instead.
reject:function () {
        if (this.options.backtrack_lexer) {
            this._backtrack = true;
        } else {
            return this.parseError('Lexical error on line ' + (this.yylineno + 1) + '. You can only invoke reject() in the lexer when the lexer is of the backtracking persuasion (options.backtrack_lexer = true).\n' + this.showPosition(), {
                text: "",
                token: null,
                line: this.yylineno
            });

        }
        return this;
    },

// retain first n characters of the match
less:function (n) {
        this.unput(this.match.slice(n));
    },

// displays already matched input, i.e. for error messages
pastInput:function () {
        var past = this.matched.substr(0, this.matched.length - this.match.length);
        return (past.length > 20 ? '...':'') + past.substr(-20).replace(/\n/g, "");
    },

// displays upcoming input, i.e. for error messages
upcomingInput:function () {
        var next = this.match;
        if (next.length < 20) {
            next += this._input.substr(0, 20-next.length);
        }
        return (next.substr(0,20) + (next.length > 20 ? '...' : '')).replace(/\n/g, "");
    },

// displays the character position where the lexing error occurred, i.e. for error messages
showPosition:function () {
        var pre = this.pastInput();
        var c = new Array(pre.length + 1).join("-");
        return pre + this.upcomingInput() + "\n" + c + "^";
    },

// test the lexed token: return FALSE when not a match, otherwise return token
test_match:function (match, indexed_rule) {
        var token,
            lines,
            backup;

        if (this.options.backtrack_lexer) {
            // save context
            backup = {
                yylineno: this.yylineno,
                yylloc: {
                    first_line: this.yylloc.first_line,
                    last_line: this.last_line,
                    first_column: this.yylloc.first_column,
                    last_column: this.yylloc.last_column
                },
                yytext: this.yytext,
                match: this.match,
                matches: this.matches,
                matched: this.matched,
                yyleng: this.yyleng,
                offset: this.offset,
                _more: this._more,
                _input: this._input,
                yy: this.yy,
                conditionStack: this.conditionStack.slice(0),
                done: this.done
            };
            if (this.options.ranges) {
                backup.yylloc.range = this.yylloc.range.slice(0);
            }
        }

        lines = match[0].match(/(?:\r\n?|\n).*/g);
        if (lines) {
            this.yylineno += lines.length;
        }
        this.yylloc = {
            first_line: this.yylloc.last_line,
            last_line: this.yylineno + 1,
            first_column: this.yylloc.last_column,
            last_column: lines ?
                         lines[lines.length - 1].length - lines[lines.length - 1].match(/\r?\n?/)[0].length :
                         this.yylloc.last_column + match[0].length
        };
        this.yytext += match[0];
        this.match += match[0];
        this.matches = match;
        this.yyleng = this.yytext.length;
        if (this.options.ranges) {
            this.yylloc.range = [this.offset, this.offset += this.yyleng];
        }
        this._more = false;
        this._backtrack = false;
        this._input = this._input.slice(match[0].length);
        this.matched += match[0];
        token = this.performAction.call(this, this.yy, this, indexed_rule, this.conditionStack[this.conditionStack.length - 1]);
        if (this.done && this._input) {
            this.done = false;
        }
        if (token) {
            return token;
        } else if (this._backtrack) {
            // recover context
            for (var k in backup) {
                this[k] = backup[k];
            }
            return false; // rule action called reject() implying the next rule should be tested instead.
        }
        return false;
    },

// return next match in input
next:function () {
        if (this.done) {
            return this.EOF;
        }
        if (!this._input) {
            this.done = true;
        }

        var token,
            match,
            tempMatch,
            index;
        if (!this._more) {
            this.yytext = '';
            this.match = '';
        }
        var rules = this._currentRules();
        for (var i = 0; i < rules.length; i++) {
            tempMatch = this._input.match(this.rules[rules[i]]);
            if (tempMatch && (!match || tempMatch[0].length > match[0].length)) {
                match = tempMatch;
                index = i;
                if (this.options.backtrack_lexer) {
                    token = this.test_match(tempMatch, rules[i]);
                    if (token !== false) {
                        return token;
                    } else if (this._backtrack) {
                        match = false;
                        continue; // rule action called reject() implying a rule MISmatch.
                    } else {
                        // else: this is a lexer rule which consumes input without producing a token (e.g. whitespace)
                        return false;
                    }
                } else if (!this.options.flex) {
                    break;
                }
            }
        }
        if (match) {
            token = this.test_match(match, rules[index]);
            if (token !== false) {
                return token;
            }
            // else: this is a lexer rule which consumes input without producing a token (e.g. whitespace)
            return false;
        }
        if (this._input === "") {
            return this.EOF;
        } else {
            return this.parseError('Lexical error on line ' + (this.yylineno + 1) + '. Unrecognized text.\n' + this.showPosition(), {
                text: "",
                token: null,
                line: this.yylineno
            });
        }
    },

// return next match that has a token
lex:function lex() {
        var r = this.next();
        if (r) {
            return r;
        } else {
            return this.lex();
        }
    },

// activates a new lexer condition state (pushes the new lexer condition state onto the condition stack)
begin:function begin(condition) {
        this.conditionStack.push(condition);
    },

// pop the previously active lexer condition state off the condition stack
popState:function popState() {
        var n = this.conditionStack.length - 1;
        if (n > 0) {
            return this.conditionStack.pop();
        } else {
            return this.conditionStack[0];
        }
    },

// produce the lexer rule set which is active for the currently active lexer condition state
_currentRules:function _currentRules() {
        if (this.conditionStack.length && this.conditionStack[this.conditionStack.length - 1]) {
            return this.conditions[this.conditionStack[this.conditionStack.length - 1]].rules;
        } else {
            return this.conditions["INITIAL"].rules;
        }
    },

// return the currently active lexer condition state; when an index argument is provided it produces the N-th previous condition state, if available
topState:function topState(n) {
        n = this.conditionStack.length - 1 - Math.abs(n || 0);
        if (n >= 0) {
            return this.conditionStack[n];
        } else {
            return "INITIAL";
        }
    },

// alias for begin(condition)
pushState:function pushState(condition) {
        this.begin(condition);
    },

// return the number of states currently on the stack
stateStackSize:function stateStackSize() {
        return this.conditionStack.length;
    },
options: [],
performAction: function anonymous(yy,yy_,$avoiding_name_collisions,YY_START) {

var YYSTATE=YY_START;
switch($avoiding_name_collisions) {
case 0:this.pushState("INSIDE_PARENS"); return "(";
break;
case 1:this.popState(); return ")";
break;
case 2:this.pushState("INITIAL"); return "{";
break;
case 3:this.popState(); return "}";
break;
case 4:this.pushState("INITIAL"); return "[";
break;
case 5:this.popState(); return "]";
break;
case 6:return "|";
break;
case 7:return "->";
break;
case 8:return YY_START === "INITIAL" ? "NEWLINE" : null;
break;
case 9:/* skip other whitespace */
break;
case 10:return YY_START === "INITIAL" ? "NEWLINE" : null; /* skip comments */
break;
case 11:return "SIGN";
break;
case 12:return "AND";
break;
case 13:return "OR";
break;
case 14:return "NEG_NUMBER";
break;
case 15:return "NEG_NUMBER";
break;
case 16:return "NUMBER";
break;
case 17:return "NUMBER";
break;
case 18:return "REGEX";
break;
case 19:return "*";
break;
case 20:return "/";
break;
case 21:return "-";
break;
case 22:return "+";
break;
case 23:return "%";
break;
case 24:return "@";
break;
case 25:return "?";
break;
case 26:return "=";
break;
case 27:return "!";
break;
case 28:return "TYPEDELIM";
break;
case 29:return ":";
break;
case 30:return "DOT";
break;
case 31:return ",";
break;
case 32:return "STRING";
break;
case 33:return "STRING";
break;
case 34:return "IDENTIFIER";
break;
case 35:return "JAVASCRIPT";
break;
case 36:return "EOF";
break;
case 37:return "INVALID";
break;
}
},
rules: [/^(?:\()/,/^(?:\))/,/^(?:\{)/,/^(?:\})/,/^(?:\[)/,/^(?:\])/,/^(?:\|)/,/^(?:->)/,/^(?:(\r\n?|\n))/,/^(?:[^\S\r\n]+)/,/^(?:\/\/.*?(\r\n?|\n))/,/^(?:==|!=|<=|>=|<|>)/,/^(?:and)/,/^(?:or)/,/^(?:-[0-9]+)/,/^(?:-([0-9]+)?\.[0-9]+)/,/^(?:[0-9]+)/,/^(?:([0-9]+)?\.[0-9]+)/,/^(?:\/(\\.|[^\/\n])*\/[gim]*)/,/^(?:\*)/,/^(?:\/)/,/^(?:-)/,/^(?:\+)/,/^(?:\%)/,/^(?:@)/,/^(?:\?)/,/^(?:=)/,/^(?:\!)/,/^(?:\:\:)/,/^(?:\:)/,/^(?:\.)/,/^(?:,)/,/^(?:\"(\\.|[^"\n])*?\")/,/^(?:\'(\\.|[^'\n])*?\')/,/^(?:[a-zA-Z_$][a-zA-Z0-9_$]*)/,/^(?:```.*?```)/,/^(?:$)/,/^(?:.)/],
conditions: {"INSIDE_PARENS":{"rules":[0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37],"inclusive":true},"INITIAL":{"rules":[0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37],"inclusive":true}}
};
return lexer;
})();
parser.lexer = lexer;
function Parser () {
  this.yy = {};
}
Parser.prototype = parser;parser.Parser = Parser;
return new Parser;
})();

parser.yy = require('./parse-tree.js');
module.exports = parser;
