// aletheia/lib/aletheia/parser.js
/* parser generated by jison 0.4.13 */
/*
  Returns a Parser object of the following structure:

  Parser: {
    yy: {}
  }

  Parser.prototype: {
    yy: {},
    trace: function(),
    symbols_: {associative list: name ==> number},
    terminals_: {associative list: number ==> name},
    productions_: [...],
    performAction: function anonymous(yytext, yyleng, yylineno, yy, yystate, $$, _$),
    table: [...],
    defaultActions: {...},
    parseError: function(str, hash),
    parse: function(input),

    lexer: {
        EOF: 1,
        parseError: function(str, hash),
        setInput: function(input),
        input: function(),
        unput: function(str),
        more: function(),
        less: function(n),
        pastInput: function(),
        upcomingInput: function(),
        showPosition: function(),
        test_match: function(regex_match_array, rule_index),
        next: function(),
        lex: function(),
        begin: function(condition),
        popState: function(),
        _currentRules: function(),
        topState: function(),
        pushState: function(condition),

        options: {
            ranges: boolean           (optional: true ==> token location info will include a .range[] member)
            flex: boolean             (optional: true ==> flex-like lexing behaviour where the rules are tested exhaustively to find the longest match)
            backtrack_lexer: boolean  (optional: true ==> lexer regexes are tested in order and for each matching regex the action code is invoked; the lexer terminates the scan when a token is returned by the action code)
        },

        performAction: function(yy, yy_, $avoiding_name_collisions, YY_START),
        rules: [...],
        conditions: {associative list: name ==> set},
    }
  }


  token location info (@$, _$, etc.): {
    first_line: n,
    last_line: n,
    first_column: n,
    last_column: n,
    range: [start_number, end_number]       (where the numbers are indexes into the input string, regular zero-based)
  }


  the parseError function receives a 'hash' object with these members for lexer and parser errors: {
    text:        (matched text)
    token:       (the produced terminal token, if any)
    line:        (yylineno)
  }
  while parser (grammar) errors will also provide these members, i.e. parser errors deliver a superset of attributes: {
    loc:         (yylloc)
    expected:    (string describing the set of expected tokens)
    recoverable: (boolean: TRUE when the parser has a error recovery rule available for this particular error)
  }
*/
var parser = (function(){
var parser = {trace: function trace() {
        Jison.print.apply(null, arguments);
    },
yy: {},
symbols_: {"error":2,"program":3,"statementList":4,"EOF":5,"statementListBody":6,"separator":7,"statement":8,"NEWLINE":9,"lvalue":10,"=":11,"expression":12,"unitList":13,"additive":14,"booleanOp":15,"comparison":16,"booleanAndOp":17,"booleanOrOp":18,"AND":19,"unitExpression":20,"OR":21,"SIGN":22,"(":23,")":24,"function":25,"literal":26,"singleUnitList":27,"JAVASCRIPT":28,"IDENTIFIER":29,"tableaccess":30,"DOT":31,"@":32,"tableNameLiteral":33,"table":34,"REGEX":35,"NUMBER":36,"NEG_NUMBER":37,"STRING":38,"{":39,"}":40,":":41,"fieldList":42,"fieldListBody":43,"field":44,",":45,"[":46,"]":47,"|":48,"multiplicative":49,"+":50,"-":51,"negative":52,"*":53,"/":54,"%":55,"$accept":0,"$end":1},
terminals_: {2:"error",5:"EOF",9:"NEWLINE",11:"=",19:"AND",21:"OR",22:"SIGN",23:"(",24:")",28:"JAVASCRIPT",29:"IDENTIFIER",31:"DOT",32:"@",35:"REGEX",36:"NUMBER",37:"NEG_NUMBER",38:"STRING",39:"{",40:"}",41:":",45:",",46:"[",47:"]",48:"|",50:"+",51:"-",53:"*",54:"/",55:"%"},
productions_: [0,[3,2],[3,1],[4,1],[4,2],[4,2],[4,3],[6,1],[6,3],[7,1],[7,2],[8,3],[8,3],[8,1],[12,1],[12,1],[12,1],[15,1],[15,1],[15,1],[17,3],[17,3],[17,3],[17,3],[17,3],[17,3],[18,3],[18,3],[18,3],[18,3],[18,3],[18,3],[16,3],[20,3],[20,1],[20,1],[20,1],[20,1],[20,1],[10,1],[10,1],[13,2],[13,2],[27,3],[30,3],[30,3],[26,1],[26,1],[26,1],[33,1],[33,1],[33,1],[34,2],[34,3],[34,3],[34,3],[42,1],[42,2],[42,2],[42,3],[43,1],[43,3],[43,3],[43,4],[44,1],[44,3],[44,3],[44,3],[44,3],[25,3],[25,5],[25,5],[25,5],[25,7],[25,7],[14,1],[14,3],[14,3],[49,1],[49,3],[49,3],[49,3],[52,2],[52,1]],
performAction: function anonymous(yytext, yyleng, yylineno, yy, yystate /* action[1] */, $$ /* vstack */, _$ /* lstack */) {
/* this == yyval */

var $0 = $$.length - 1;
switch (yystate) {
case 1:return $$[$0-1];
break;
case 2:return [];
break;
case 3:this.$ = $$[$0];
break;
case 4:this.$ = $$[$0-1];
break;
case 5:this.$ = $$[$0];
break;
case 6:this.$ = $$[$0-1];
break;
case 7:this.$ = [$$[$0]];
break;
case 8:this.$ = $$[$0-2]; $$[$0-2].push($$[$0]);
break;
case 9:
break;
case 10:
break;
case 11:this.$ = yy.Declaration($$[$0-2], $$[$0]);
break;
case 12:this.$ = yy.Assignment($$[$0-2], $$[$0]);
break;
case 13:this.$ = $$[$0];
break;
case 14:this.$ = $$[$0];
break;
case 15:this.$ = $$[$0];
break;
case 16:this.$ = $$[$0]
break;
case 17:this.$ = $$[$0]
break;
case 18:this.$ = $$[$0]
break;
case 19:this.$ = $$[$0]
break;
case 20:this.$ = yy.Operation($$[$0-2], $$[$0-1], $$[$0]);
break;
case 21:this.$ = yy.Operation($$[$0-2], $$[$0-1], $$[$0]);
break;
case 22:this.$ = yy.Operation($$[$0-2], $$[$0-1], $$[$0]);
break;
case 23:this.$ = yy.Operation($$[$0-2], $$[$0-1], $$[$0]);
break;
case 24:this.$ = yy.Operation($$[$0-2], $$[$0-1], $$[$0]);
break;
case 25:this.$ = yy.Operation($$[$0-2], $$[$0-1], $$[$0]);
break;
case 26:this.$ = yy.Operation($$[$0-2], $$[$0-1], $$[$0]);
break;
case 27:this.$ = yy.Operation($$[$0-2], $$[$0-1], $$[$0]);
break;
case 28:this.$ = yy.Operation($$[$0-2], $$[$0-1], $$[$0]);
break;
case 29:this.$ = yy.Operation($$[$0-2], $$[$0-1], $$[$0]);
break;
case 30:this.$ = yy.Operation($$[$0-2], $$[$0-1], $$[$0]);
break;
case 31:this.$ = yy.Operation($$[$0-2], $$[$0-1], $$[$0]);
break;
case 32:this.$ = yy.Operation($$[$0-2], $$[$0-1], $$[$0]);
break;
case 33:this.$ = $$[$0-1];
break;
case 34:this.$ = $$[$0];
break;
case 35:this.$ = $$[$0];
break;
case 36:this.$ = $$[$0];
break;
case 37:this.$ = $$[$0];
break;
case 38:this.$ = yy.Javascript($$[$0].slice(3, -3));
break;
case 39:this.$ = yy.Variable($$[$0]);
break;
case 40:this.$ = $$[$0];
break;
case 41:this.$ = yy.UnitList([$$[$0-1], $$[$0]]);
break;
case 42:this.$ = $$[$0]; $$[$0].units.unshift($$[$0-1]);
break;
case 43:this.$ = yy.UnitList([$$[$0-2]]);
break;
case 44:this.$ = new yy.TableAccess($$[$0-2], $$[$0]);
break;
case 45:this.$ = new yy.TableAccess($$[$0-2], $$[$0]);
break;
case 46:this.$ = $$[$0];
break;
case 47:this.$ = $$[$0];
break;
case 48:this.$ = yy.Regex($$[$0]);
break;
case 49:this.$ = Number($$[$0]);
break;
case 50:this.$ = Number($$[$0]);
break;
case 51:this.$ = yy.String($$[$0]);
break;
case 52:this.$ = new yy.Table([], false);
break;
case 53:this.$ = new yy.Table([], true);
break;
case 54:this.$ = new yy.Table([], true);
break;
case 55:this.$ = new yy.Table($$[$0-1], false);
break;
case 56:this.$ = $$[$0];
break;
case 57:this.$ = $$[$0-1];
break;
case 58:this.$ = $$[$0];
break;
case 59:this.$ = $$[$0-1];
break;
case 60:this.$ = [$$[$0]];
break;
case 61:this.$ = $$[$0-2]; $$[$0-2].push($$[$0]);
break;
case 62:this.$ = $$[$0-2]; $$[$0-2].push($$[$0]);
break;
case 63:this.$ = $$[$0-3]; $$[$0-3].push($$[$0]);
break;
case 64:this.$ = yy.Field(null, $$[$0]);
break;
case 65:this.$ = yy.Field($$[$0-2], $$[$0]);
break;
case 66:this.$ = yy.Field($$[$0-2], $$[$0]);
break;
case 67:this.$ = yy.Field($$[$0-2], $$[$0]);
break;
case 68:this.$ = yy.Field($$[$0-2], $$[$0]);
break;
case 69:this.$ = yy.RetLambda([], $$[$0-1]);
break;
case 70:this.$ = yy.RetLambda($$[$0-3].units, $$[$0-1]);
break;
case 71:this.$ = yy.RetLambda([$$[$0-3]], $$[$0-1]);
break;
case 72:this.$ = yy.Lambda([], $$[$0-2]);
break;
case 73:this.$ = yy.Lambda($$[$0-5].units, $$[$0-2]);
break;
case 74:this.$ = yy.Lambda([$$[$0-5]], $$[$0-2]);
break;
case 75:this.$ = $$[$0];
break;
case 76:this.$ = yy.Operation($$[$0-2], $$[$0-1], $$[$0]);
break;
case 77:this.$ = yy.Operation($$[$0-2], $$[$0-1], $$[$0]);
break;
case 78:this.$ = $$[$0];
break;
case 79:this.$ = yy.Operation($$[$0-2], $$[$0-1], $$[$0]);
break;
case 80:this.$ = yy.Operation($$[$0-2], $$[$0-1], $$[$0]);
break;
case 81:this.$ = yy.Operation($$[$0-2], $$[$0-1], $$[$0]);
break;
case 82:this.$ = yy.Operation(null, $$[$0-1], $$[$0]);
break;
case 83:this.$ = $$[$0];
break;
}
},
table: [{3:1,4:2,5:[1,3],6:4,7:5,8:6,9:[1,7],10:8,12:10,13:9,14:14,15:15,16:22,17:23,18:24,20:13,23:[1,16],25:17,26:18,27:19,28:[1,20],29:[1,11],30:12,33:26,34:27,35:[1,28],36:[1,30],37:[1,31],38:[1,32],39:[1,33],46:[1,25],49:21,51:[1,34],52:29},{1:[3]},{5:[1,35]},{1:[2,2]},{5:[2,3],7:36,9:[1,7]},{6:37,8:6,9:[1,38],10:8,12:10,13:9,14:14,15:15,16:22,17:23,18:24,20:13,23:[1,16],25:17,26:18,27:19,28:[1,20],29:[1,11],30:12,33:26,34:27,35:[1,28],36:[1,30],37:[1,31],38:[1,32],39:[1,33],46:[1,25],49:21,51:[1,34],52:29},{5:[2,7],9:[2,7]},{5:[2,9],9:[2,9],23:[2,9],28:[2,9],29:[2,9],35:[2,9],36:[2,9],37:[2,9],38:[2,9],39:[2,9],40:[2,9],46:[2,9],47:[2,9],51:[2,9]},{5:[2,36],9:[2,36],11:[1,39],19:[2,36],21:[2,36],22:[2,36],23:[2,36],28:[2,36],29:[2,36],31:[2,36],32:[2,36],35:[2,36],36:[2,36],37:[2,36],38:[2,36],39:[2,36],46:[2,36],47:[2,36],48:[2,36],50:[2,36],51:[2,36],53:[2,36],54:[2,36],55:[2,36]},{5:[2,14],9:[2,14],11:[1,40],47:[2,14]},{5:[2,13],9:[2,13],47:[2,13]},{5:[2,39],9:[2,39],11:[2,39],19:[2,39],21:[2,39],22:[2,39],23:[2,39],24:[2,39],28:[2,39],29:[2,39],31:[2,39],32:[2,39],35:[2,39],36:[2,39],37:[2,39],38:[2,39],39:[2,39],40:[2,39],45:[2,39],46:[2,39],47:[2,39],48:[2,39],50:[2,39],51:[2,39],53:[2,39],54:[2,39],55:[2,39]},{5:[2,40],9:[2,40],11:[2,40],19:[2,40],21:[2,40],22:[2,40],23:[2,40],24:[2,40],28:[2,40],29:[2,40],31:[2,40],32:[2,40],35:[2,40],36:[2,40],37:[2,40],38:[2,40],39:[2,40],40:[2,40],45:[2,40],46:[2,40],47:[2,40],48:[2,40],50:[2,40],51:[2,40],53:[2,40],54:[2,40],55:[2,40]},{5:[2,83],9:[2,83],10:48,13:42,19:[1,46],20:41,21:[1,47],22:[2,83],23:[1,45],24:[2,83],25:17,26:18,27:19,28:[1,20],29:[1,11],30:12,31:[1,43],32:[1,44],33:26,34:27,35:[1,28],36:[1,30],37:[1,31],38:[1,32],39:[1,33],40:[2,83],45:[2,83],46:[1,25],47:[2,83],50:[2,83],51:[2,83],53:[2,83],54:[2,83],55:[2,83]},{5:[2,15],9:[2,15],22:[1,51],24:[2,15],40:[2,15],45:[2,15],47:[2,15],50:[1,49],51:[1,50]},{5:[2,16],9:[2,16],24:[2,16],40:[2,16],45:[2,16],47:[2,16]},{10:48,12:52,13:53,14:14,15:15,16:22,17:23,18:24,20:13,23:[1,16],25:17,26:18,27:19,28:[1,20],29:[1,11],30:12,33:26,34:27,35:[1,28],36:[1,30],37:[1,31],38:[1,32],39:[1,33],46:[1,25],49:21,51:[1,34],52:29},{5:[2,34],9:[2,34],11:[2,34],19:[2,34],21:[2,34],22:[2,34],23:[2,34],24:[2,34],28:[2,34],29:[2,34],31:[2,34],32:[2,34],35:[2,34],36:[2,34],37:[2,34],38:[2,34],39:[2,34],40:[2,34],45:[2,34],46:[2,34],47:[2,34],48:[2,34],50:[2,34],51:[2,34],53:[2,34],54:[2,34],55:[2,34]},{5:[2,35],9:[2,35],11:[2,35],19:[2,35],21:[2,35],22:[2,35],23:[2,35],24:[2,35],28:[2,35],29:[2,35],31:[2,35],32:[2,35],35:[2,35],36:[2,35],37:[2,35],38:[2,35],39:[2,35],40:[2,35],45:[2,35],46:[2,35],47:[2,35],48:[2,35],50:[2,35],51:[2,35],53:[2,35],54:[2,35],55:[2,35]},{5:[2,37],9:[2,37],11:[2,37],19:[2,37],21:[2,37],22:[2,37],23:[2,37],24:[2,37],28:[2,37],29:[2,37],31:[2,37],32:[2,37],35:[2,37],36:[2,37],37:[2,37],38:[2,37],39:[2,37],40:[2,37],45:[2,37],46:[2,37],47:[2,37],48:[2,37],50:[2,37],51:[2,37],53:[2,37],54:[2,37],55:[2,37]},{5:[2,38],9:[2,38],11:[2,38],19:[2,38],21:[2,38],22:[2,38],23:[2,38],24:[2,38],28:[2,38],29:[2,38],31:[2,38],32:[2,38],35:[2,38],36:[2,38],37:[2,38],38:[2,38],39:[2,38],40:[2,38],45:[2,38],46:[2,38],47:[2,38],48:[2,38],50:[2,38],51:[2,38],53:[2,38],54:[2,38],55:[2,38]},{5:[2,75],9:[2,75],19:[2,75],21:[2,75],22:[2,75],24:[2,75],40:[2,75],45:[2,75],47:[2,75],50:[2,75],51:[2,75],53:[1,54],54:[1,55],55:[1,56]},{5:[2,17],9:[2,17],19:[1,57],21:[1,58],24:[2,17],40:[2,17],45:[2,17],47:[2,17]},{5:[2,18],9:[2,18],19:[1,59],24:[2,18],40:[2,18],45:[2,18],47:[2,18]},{5:[2,19],9:[2,19],21:[1,60],24:[2,19],40:[2,19],45:[2,19],47:[2,19]},{7:64,8:61,9:[1,7],10:8,12:10,13:62,14:14,15:15,16:22,17:23,18:24,20:63,23:[1,16],25:17,26:18,27:19,28:[1,20],29:[1,11],30:12,33:26,34:27,35:[1,28],36:[1,30],37:[1,31],38:[1,32],39:[1,33],46:[1,25],49:21,51:[1,34],52:29},{5:[2,46],9:[2,46],11:[2,46],19:[2,46],21:[2,46],22:[2,46],23:[2,46],24:[2,46],28:[2,46],29:[2,46],31:[2,46],32:[2,46],35:[2,46],36:[2,46],37:[2,46],38:[2,46],39:[2,46],40:[2,46],45:[2,46],46:[2,46],47:[2,46],48:[2,46],50:[2,46],51:[2,46],53:[2,46],54:[2,46],55:[2,46]},{5:[2,47],9:[2,47],11:[2,47],19:[2,47],21:[2,47],22:[2,47],23:[2,47],24:[2,47],28:[2,47],29:[2,47],31:[2,47],32:[2,47],35:[2,47],36:[2,47],37:[2,47],38:[2,47],39:[2,47],40:[2,47],45:[2,47],46:[2,47],47:[2,47],48:[2,47],50:[2,47],51:[2,47],53:[2,47],54:[2,47],55:[2,47]},{5:[2,48],9:[2,48],11:[2,48],19:[2,48],21:[2,48],22:[2,48],23:[2,48],24:[2,48],28:[2,48],29:[2,48],31:[2,48],32:[2,48],35:[2,48],36:[2,48],37:[2,48],38:[2,48],39:[2,48],40:[2,48],45:[2,48],46:[2,48],47:[2,48],48:[2,48],50:[2,48],51:[2,48],53:[2,48],54:[2,48],55:[2,48]},{5:[2,78],9:[2,78],19:[2,78],21:[2,78],22:[2,78],24:[2,78],40:[2,78],45:[2,78],47:[2,78],50:[2,78],51:[2,78],53:[2,78],54:[2,78],55:[2,78]},{5:[2,49],9:[2,49],11:[2,49],19:[2,49],21:[2,49],22:[2,49],23:[2,49],24:[2,49],28:[2,49],29:[2,49],31:[2,49],32:[2,49],35:[2,49],36:[2,49],37:[2,49],38:[2,49],39:[2,49],40:[2,49],41:[2,49],45:[2,49],46:[2,49],47:[2,49],48:[2,49],50:[2,49],51:[2,49],53:[2,49],54:[2,49],55:[2,49]},{5:[2,50],9:[2,50],11:[2,50],19:[2,50],21:[2,50],22:[2,50],23:[2,50],24:[2,50],28:[2,50],29:[2,50],31:[2,50],32:[2,50],35:[2,50],36:[2,50],37:[2,50],38:[2,50],39:[2,50],40:[2,50],41:[2,50],45:[2,50],46:[2,50],47:[2,50],48:[2,50],50:[2,50],51:[2,50],53:[2,50],54:[2,50],55:[2,50]},{5:[2,51],9:[2,51],11:[2,51],19:[2,51],21:[2,51],22:[2,51],23:[2,51],24:[2,51],28:[2,51],29:[2,51],31:[2,51],32:[2,51],35:[2,51],36:[2,51],37:[2,51],38:[2,51],39:[2,51],40:[2,51],41:[2,51],45:[2,51],46:[2,51],47:[2,51],48:[2,51],50:[2,51],51:[2,51],53:[2,51],54:[2,51],55:[2,51]},{7:70,9:[1,7],10:48,11:[1,67],12:72,13:53,14:14,15:15,16:22,17:23,18:24,20:13,23:[1,16],25:17,26:18,27:19,28:[1,20],29:[1,73],30:12,33:74,34:27,35:[1,28],36:[1,30],37:[1,31],38:[1,32],39:[1,33],40:[1,65],41:[1,66],42:68,43:69,44:71,46:[1,25],49:21,51:[1,34],52:29},{10:48,20:76,23:[1,16],25:17,26:18,27:19,28:[1,20],29:[1,11],30:12,33:26,34:27,35:[1,28],36:[1,30],37:[1,31],38:[1,32],39:[1,33],46:[1,25],51:[1,34],52:75},{1:[2,1]},{5:[2,4],8:77,9:[1,38],10:8,12:10,13:9,14:14,15:15,16:22,17:23,18:24,20:13,23:[1,16],25:17,26:18,27:19,28:[1,20],29:[1,11],30:12,33:26,34:27,35:[1,28],36:[1,30],37:[1,31],38:[1,32],39:[1,33],46:[1,25],49:21,51:[1,34],52:29},{5:[2,5],7:78,9:[1,7]},{5:[2,10],9:[2,10],23:[2,10],28:[2,10],29:[2,10],35:[2,10],36:[2,10],37:[2,10],38:[2,10],39:[2,10],40:[2,10],46:[2,10],47:[2,10],51:[2,10]},{10:48,12:79,13:53,14:14,15:15,16:22,17:23,18:24,20:13,23:[1,16],25:17,26:18,27:19,28:[1,20],29:[1,11],30:12,33:26,34:27,35:[1,28],36:[1,30],37:[1,31],38:[1,32],39:[1,33],46:[1,25],49:21,51:[1,34],52:29},{10:48,12:80,13:53,14:14,15:15,16:22,17:23,18:24,20:13,23:[1,16],25:17,26:18,27:19,28:[1,20],29:[1,11],30:12,33:26,34:27,35:[1,28],36:[1,30],37:[1,31],38:[1,32],39:[1,33],46:[1,25],49:21,51:[1,34],52:29},{5:[2,41],9:[2,41],10:48,11:[2,41],13:42,20:41,23:[1,45],24:[2,41],25:17,26:18,27:19,28:[1,20],29:[1,11],30:12,31:[1,43],32:[1,44],33:26,34:27,35:[1,28],36:[1,30],37:[1,31],38:[1,32],39:[1,33],40:[2,41],45:[2,41],46:[1,25],47:[2,41],48:[2,41]},{5:[2,42],9:[2,42],11:[2,42],24:[2,42],40:[2,42],45:[2,42],47:[2,42],48:[2,42]},{29:[1,81]},{10:48,20:82,23:[1,16],25:17,26:18,27:19,28:[1,20],29:[1,11],30:12,33:26,34:27,35:[1,28],36:[1,30],37:[1,31],38:[1,32],39:[1,33],46:[1,25]},{10:48,12:52,13:53,14:14,15:15,16:22,17:23,18:24,20:13,23:[1,16],24:[1,83],25:17,26:18,27:19,28:[1,20],29:[1,11],30:12,33:26,34:27,35:[1,28],36:[1,30],37:[1,31],38:[1,32],39:[1,33],46:[1,25],49:21,51:[1,34],52:29},{10:48,14:86,16:84,20:85,23:[1,16],25:17,26:18,27:19,28:[1,20],29:[1,11],30:12,33:26,34:27,35:[1,28],36:[1,30],37:[1,31],38:[1,32],39:[1,33],46:[1,25],49:21,51:[1,34],52:29},{10:48,14:86,16:87,20:88,23:[1,16],25:17,26:18,27:19,28:[1,20],29:[1,11],30:12,33:26,34:27,35:[1,28],36:[1,30],37:[1,31],38:[1,32],39:[1,33],46:[1,25],49:21,51:[1,34],52:29},{5:[2,36],9:[2,36],11:[2,36],19:[2,36],21:[2,36],22:[2,36],23:[2,36],24:[2,36],28:[2,36],29:[2,36],31:[2,36],32:[2,36],35:[2,36],36:[2,36],37:[2,36],38:[2,36],39:[2,36],40:[2,36],45:[2,36],46:[2,36],47:[2,36],48:[2,36],50:[2,36],51:[2,36],53:[2,36],54:[2,36],55:[2,36]},{10:48,14:89,20:76,23:[1,16],25:17,26:18,27:19,28:[1,20],29:[1,11],30:12,33:26,34:27,35:[1,28],36:[1,30],37:[1,31],38:[1,32],39:[1,33],46:[1,25],49:21,51:[1,34],52:29},{10:48,14:90,20:76,23:[1,16],25:17,26:18,27:19,28:[1,20],29:[1,11],30:12,33:26,34:27,35:[1,28],36:[1,30],37:[1,31],38:[1,32],39:[1,33],46:[1,25],49:21,51:[1,34],52:29},{10:48,14:91,20:76,23:[1,16],25:17,26:18,27:19,28:[1,20],29:[1,11],30:12,33:26,34:27,35:[1,28],36:[1,30],37:[1,31],38:[1,32],39:[1,33],46:[1,25],49:21,51:[1,34],52:29},{24:[1,92]},{5:[2,14],9:[2,14],24:[2,14],40:[2,14],45:[2,14],47:[2,14]},{10:48,20:76,23:[1,16],25:17,26:18,27:19,28:[1,20],29:[1,11],30:12,33:26,34:27,35:[1,28],36:[1,30],37:[1,31],38:[1,32],39:[1,33],46:[1,25],49:93,51:[1,34],52:29},{10:48,20:76,23:[1,16],25:17,26:18,27:19,28:[1,20],29:[1,11],30:12,33:26,34:27,35:[1,28],36:[1,30],37:[1,31],38:[1,32],39:[1,33],46:[1,25],49:94,51:[1,34],52:29},{10:48,20:76,23:[1,16],25:17,26:18,27:19,28:[1,20],29:[1,11],30:12,33:26,34:27,35:[1,28],36:[1,30],37:[1,31],38:[1,32],39:[1,33],46:[1,25],49:95,51:[1,34],52:29},{10:48,14:86,16:96,20:97,23:[1,16],25:17,26:18,27:19,28:[1,20],29:[1,11],30:12,33:26,34:27,35:[1,28],36:[1,30],37:[1,31],38:[1,32],39:[1,33],46:[1,25],49:21,51:[1,34],52:29},{10:48,14:86,16:98,20:99,23:[1,16],25:17,26:18,27:19,28:[1,20],29:[1,11],30:12,33:26,34:27,35:[1,28],36:[1,30],37:[1,31],38:[1,32],39:[1,33],46:[1,25],49:21,51:[1,34],52:29},{10:48,14:86,16:100,20:101,23:[1,16],25:17,26:18,27:19,28:[1,20],29:[1,11],30:12,33:26,34:27,35:[1,28],36:[1,30],37:[1,31],38:[1,32],39:[1,33],46:[1,25],49:21,51:[1,34],52:29},{10:48,14:86,16:102,20:103,23:[1,16],25:17,26:18,27:19,28:[1,20],29:[1,11],30:12,33:26,34:27,35:[1,28],36:[1,30],37:[1,31],38:[1,32],39:[1,33],46:[1,25],49:21,51:[1,34],52:29},{47:[1,104]},{11:[1,40],47:[2,14],48:[1,105]},{10:48,13:42,19:[1,46],20:41,21:[1,47],22:[2,83],23:[1,45],25:17,26:18,27:19,28:[1,20],29:[1,11],30:12,31:[1,43],32:[1,44],33:26,34:27,35:[1,28],36:[1,30],37:[1,31],38:[1,32],39:[1,33],46:[1,25],47:[2,83],48:[1,106],50:[2,83],51:[2,83],53:[2,83],54:[2,83],55:[2,83]},{6:107,8:6,9:[1,38],10:8,12:10,13:9,14:14,15:15,16:22,17:23,18:24,20:13,23:[1,16],25:17,26:18,27:19,28:[1,20],29:[1,11],30:12,33:26,34:27,35:[1,28],36:[1,30],37:[1,31],38:[1,32],39:[1,33],46:[1,25],49:21,51:[1,34],52:29},{5:[2,52],9:[2,52],11:[2,52],19:[2,52],21:[2,52],22:[2,52],23:[2,52],24:[2,52],28:[2,52],29:[2,52],31:[2,52],32:[2,52],35:[2,52],36:[2,52],37:[2,52],38:[2,52],39:[2,52],40:[2,52],45:[2,52],46:[2,52],47:[2,52],48:[2,52],50:[2,52],51:[2,52],53:[2,52],54:[2,52],55:[2,52]},{40:[1,108]},{40:[1,109]},{40:[1,110]},{7:111,9:[1,7],40:[2,56],45:[1,112]},{9:[1,38],10:48,12:72,13:53,14:14,15:15,16:22,17:23,18:24,20:13,23:[1,16],25:17,26:18,27:19,28:[1,20],29:[1,73],30:12,33:74,34:27,35:[1,28],36:[1,30],37:[1,31],38:[1,32],39:[1,33],43:113,44:71,46:[1,25],49:21,51:[1,34],52:29},{9:[2,60],40:[2,60],45:[2,60]},{9:[2,64],40:[2,64],45:[2,64]},{9:[2,39],11:[1,115],19:[2,39],21:[2,39],22:[2,39],23:[2,39],28:[2,39],29:[2,39],31:[2,39],32:[2,39],35:[2,39],36:[2,39],37:[2,39],38:[2,39],39:[2,39],40:[2,39],41:[1,114],45:[2,39],46:[2,39],50:[2,39],51:[2,39],53:[2,39],54:[2,39],55:[2,39]},{9:[2,46],11:[1,117],19:[2,46],21:[2,46],22:[2,46],23:[2,46],28:[2,46],29:[2,46],31:[2,46],32:[2,46],35:[2,46],36:[2,46],37:[2,46],38:[2,46],39:[2,46],40:[2,46],41:[1,116],45:[2,46],46:[2,46],50:[2,46],51:[2,46],53:[2,46],54:[2,46],55:[2,46]},{5:[2,82],9:[2,82],19:[2,82],21:[2,82],22:[2,82],24:[2,82],40:[2,82],45:[2,82],47:[2,82],50:[2,82],51:[2,82],53:[2,82],54:[2,82],55:[2,82]},{5:[2,83],9:[2,83],19:[2,83],21:[2,83],22:[2,83],23:[1,118],24:[2,83],31:[1,43],32:[1,44],40:[2,83],45:[2,83],47:[2,83],50:[2,83],51:[2,83],53:[2,83],54:[2,83],55:[2,83]},{5:[2,8],9:[2,8]},{5:[2,6],8:77,9:[1,38],10:8,12:10,13:9,14:14,15:15,16:22,17:23,18:24,20:13,23:[1,16],25:17,26:18,27:19,28:[1,20],29:[1,11],30:12,33:26,34:27,35:[1,28],36:[1,30],37:[1,31],38:[1,32],39:[1,33],46:[1,25],49:21,51:[1,34],52:29},{5:[2,11],9:[2,11],47:[2,11]},{5:[2,12],9:[2,12],47:[2,12]},{5:[2,44],9:[2,44],11:[2,44],19:[2,44],21:[2,44],22:[2,44],23:[2,44],24:[2,44],28:[2,44],29:[2,44],31:[2,44],32:[2,44],35:[2,44],36:[2,44],37:[2,44],38:[2,44],39:[2,44],40:[2,44],45:[2,44],46:[2,44],47:[2,44],48:[2,44],50:[2,44],51:[2,44],53:[2,44],54:[2,44],55:[2,44]},{5:[2,45],9:[2,45],11:[2,45],19:[2,45],21:[2,45],22:[2,45],23:[1,118],24:[2,45],28:[2,45],29:[2,45],31:[2,45],32:[2,45],35:[2,45],36:[2,45],37:[2,45],38:[2,45],39:[2,45],40:[2,45],45:[2,45],46:[2,45],47:[2,45],48:[2,45],50:[2,45],51:[2,45],53:[2,45],54:[2,45],55:[2,45]},{5:[2,43],9:[2,43],11:[2,43],19:[2,43],21:[2,43],22:[2,43],23:[2,43],24:[2,43],28:[2,43],29:[2,43],31:[2,43],32:[2,43],35:[2,43],36:[2,43],37:[2,43],38:[2,43],39:[2,43],40:[2,43],45:[2,43],46:[2,43],47:[2,43],48:[2,43],50:[2,43],51:[2,43],53:[2,43],54:[2,43],55:[2,43]},{5:[2,22],9:[2,22],19:[2,22],24:[2,22],40:[2,22],45:[2,22],47:[2,22]},{5:[2,23],9:[2,23],19:[2,23],22:[2,83],23:[1,118],24:[2,23],31:[1,43],32:[1,44],40:[2,23],45:[2,23],47:[2,23],50:[2,83],51:[2,83],53:[2,83],54:[2,83],55:[2,83]},{22:[1,51],50:[1,49],51:[1,50]},{5:[2,28],9:[2,28],21:[2,28],24:[2,28],40:[2,28],45:[2,28],47:[2,28]},{5:[2,29],9:[2,29],21:[2,29],22:[2,83],23:[1,118],24:[2,29],31:[1,43],32:[1,44],40:[2,29],45:[2,29],47:[2,29],50:[2,83],51:[2,83],53:[2,83],54:[2,83],55:[2,83]},{5:[2,76],9:[2,76],19:[2,76],21:[2,76],22:[2,76],24:[2,76],40:[2,76],45:[2,76],47:[2,76],50:[2,76],51:[2,76]},{5:[2,77],9:[2,77],19:[2,77],21:[2,77],22:[2,77],24:[2,77],40:[2,77],45:[2,77],47:[2,77],50:[2,77],51:[2,77]},{5:[2,32],9:[2,32],19:[2,32],21:[2,32],24:[2,32],40:[2,32],45:[2,32],47:[2,32],50:[1,49],51:[1,50]},{5:[2,33],9:[2,33],11:[2,33],19:[2,33],21:[2,33],22:[2,33],23:[2,33],24:[2,33],28:[2,33],29:[2,33],31:[2,33],32:[2,33],35:[2,33],36:[2,33],37:[2,33],38:[2,33],39:[2,33],40:[2,33],45:[2,33],46:[2,33],47:[2,33],48:[2,33],50:[2,33],51:[2,33],53:[2,33],54:[2,33],55:[2,33]},{5:[2,79],9:[2,79],19:[2,79],21:[2,79],22:[2,79],24:[2,79],40:[2,79],45:[2,79],47:[2,79],50:[2,79],51:[2,79],53:[2,79],54:[2,79],55:[2,79]},{5:[2,80],9:[2,80],19:[2,80],21:[2,80],22:[2,80],24:[2,80],40:[2,80],45:[2,80],47:[2,80],50:[2,80],51:[2,80],53:[2,80],54:[2,80],55:[2,80]},{5:[2,81],9:[2,81],19:[2,81],21:[2,81],22:[2,81],24:[2,81],40:[2,81],45:[2,81],47:[2,81],50:[2,81],51:[2,81],53:[2,81],54:[2,81],55:[2,81]},{5:[2,20],9:[2,20],19:[2,20],24:[2,20],40:[2,20],45:[2,20],47:[2,20]},{5:[2,21],9:[2,21],19:[2,21],22:[2,83],23:[1,118],24:[2,21],31:[1,43],32:[1,44],40:[2,21],45:[2,21],47:[2,21],50:[2,83],51:[2,83],53:[2,83],54:[2,83],55:[2,83]},{5:[2,26],9:[2,26],21:[2,26],24:[2,26],40:[2,26],45:[2,26],47:[2,26]},{5:[2,27],9:[2,27],21:[2,27],22:[2,83],23:[1,118],24:[2,27],31:[1,43],32:[1,44],40:[2,27],45:[2,27],47:[2,27],50:[2,83],51:[2,83],53:[2,83],54:[2,83],55:[2,83]},{5:[2,24],9:[2,24],19:[2,24],24:[2,24],40:[2,24],45:[2,24],47:[2,24]},{5:[2,25],9:[2,25],19:[2,25],22:[2,83],23:[1,118],24:[2,25],31:[1,43],32:[1,44],40:[2,25],45:[2,25],47:[2,25],50:[2,83],51:[2,83],53:[2,83],54:[2,83],55:[2,83]},{5:[2,30],9:[2,30],21:[2,30],24:[2,30],40:[2,30],45:[2,30],47:[2,30]},{5:[2,31],9:[2,31],21:[2,31],22:[2,83],23:[1,118],24:[2,31],31:[1,43],32:[1,44],40:[2,31],45:[2,31],47:[2,31],50:[2,83],51:[2,83],53:[2,83],54:[2,83],55:[2,83]},{5:[2,69],9:[2,69],11:[2,69],19:[2,69],21:[2,69],22:[2,69],23:[2,69],24:[2,69],28:[2,69],29:[2,69],31:[2,69],32:[2,69],35:[2,69],36:[2,69],37:[2,69],38:[2,69],39:[2,69],40:[2,69],45:[2,69],46:[2,69],47:[2,69],48:[2,69],50:[2,69],51:[2,69],53:[2,69],54:[2,69],55:[2,69]},{7:120,8:119,9:[1,7],10:8,12:10,13:9,14:14,15:15,16:22,17:23,18:24,20:13,23:[1,16],25:17,26:18,27:19,28:[1,20],29:[1,11],30:12,33:26,34:27,35:[1,28],36:[1,30],37:[1,31],38:[1,32],39:[1,33],46:[1,25],49:21,51:[1,34],52:29},{7:122,8:121,9:[1,7],10:8,12:10,13:9,14:14,15:15,16:22,17:23,18:24,20:13,23:[1,16],25:17,26:18,27:19,28:[1,20],29:[1,11],30:12,33:26,34:27,35:[1,28],36:[1,30],37:[1,31],38:[1,32],39:[1,33],46:[1,25],49:21,51:[1,34],52:29},{7:123,9:[1,7]},{5:[2,53],9:[2,53],11:[2,53],19:[2,53],21:[2,53],22:[2,53],23:[2,53],24:[2,53],28:[2,53],29:[2,53],31:[2,53],32:[2,53],35:[2,53],36:[2,53],37:[2,53],38:[2,53],39:[2,53],40:[2,53],45:[2,53],46:[2,53],47:[2,53],48:[2,53],50:[2,53],51:[2,53],53:[2,53],54:[2,53],55:[2,53]},{5:[2,54],9:[2,54],11:[2,54],19:[2,54],21:[2,54],22:[2,54],23:[2,54],24:[2,54],28:[2,54],29:[2,54],31:[2,54],32:[2,54],35:[2,54],36:[2,54],37:[2,54],38:[2,54],39:[2,54],40:[2,54],45:[2,54],46:[2,54],47:[2,54],48:[2,54],50:[2,54],51:[2,54],53:[2,54],54:[2,54],55:[2,54]},{5:[2,55],9:[2,55],11:[2,55],19:[2,55],21:[2,55],22:[2,55],23:[2,55],24:[2,55],28:[2,55],29:[2,55],31:[2,55],32:[2,55],35:[2,55],36:[2,55],37:[2,55],38:[2,55],39:[2,55],40:[2,55],45:[2,55],46:[2,55],47:[2,55],48:[2,55],50:[2,55],51:[2,55],53:[2,55],54:[2,55],55:[2,55]},{9:[1,38],10:48,12:72,13:53,14:14,15:15,16:22,17:23,18:24,20:13,23:[1,16],25:17,26:18,27:19,28:[1,20],29:[1,73],30:12,33:74,34:27,35:[1,28],36:[1,30],37:[1,31],38:[1,32],39:[1,33],40:[2,57],44:124,46:[1,25],49:21,51:[1,34],52:29},{7:126,9:[1,7],10:48,12:72,13:53,14:14,15:15,16:22,17:23,18:24,20:13,23:[1,16],25:17,26:18,27:19,28:[1,20],29:[1,73],30:12,33:74,34:27,35:[1,28],36:[1,30],37:[1,31],38:[1,32],39:[1,33],44:125,46:[1,25],49:21,51:[1,34],52:29},{7:127,9:[1,7],40:[2,58],45:[1,112]},{10:48,12:128,13:53,14:14,15:15,16:22,17:23,18:24,20:13,23:[1,16],25:17,26:18,27:19,28:[1,20],29:[1,11],30:12,33:26,34:27,35:[1,28],36:[1,30],37:[1,31],38:[1,32],39:[1,33],46:[1,25],49:21,51:[1,34],52:29},{10:48,12:129,13:53,14:14,15:15,16:22,17:23,18:24,20:13,23:[1,16],25:17,26:18,27:19,28:[1,20],29:[1,11],30:12,33:26,34:27,35:[1,28],36:[1,30],37:[1,31],38:[1,32],39:[1,33],46:[1,25],49:21,51:[1,34],52:29},{10:48,12:130,13:53,14:14,15:15,16:22,17:23,18:24,20:13,23:[1,16],25:17,26:18,27:19,28:[1,20],29:[1,11],30:12,33:26,34:27,35:[1,28],36:[1,30],37:[1,31],38:[1,32],39:[1,33],46:[1,25],49:21,51:[1,34],52:29},{10:48,12:131,13:53,14:14,15:15,16:22,17:23,18:24,20:13,23:[1,16],25:17,26:18,27:19,28:[1,20],29:[1,11],30:12,33:26,34:27,35:[1,28],36:[1,30],37:[1,31],38:[1,32],39:[1,33],46:[1,25],49:21,51:[1,34],52:29},{24:[1,83]},{47:[1,132]},{6:133,8:6,9:[1,38],10:8,12:10,13:9,14:14,15:15,16:22,17:23,18:24,20:13,23:[1,16],25:17,26:18,27:19,28:[1,20],29:[1,11],30:12,33:26,34:27,35:[1,28],36:[1,30],37:[1,31],38:[1,32],39:[1,33],46:[1,25],49:21,51:[1,34],52:29},{47:[1,134]},{6:135,8:6,9:[1,38],10:8,12:10,13:9,14:14,15:15,16:22,17:23,18:24,20:13,23:[1,16],25:17,26:18,27:19,28:[1,20],29:[1,11],30:12,33:26,34:27,35:[1,28],36:[1,30],37:[1,31],38:[1,32],39:[1,33],46:[1,25],49:21,51:[1,34],52:29},{8:77,9:[1,38],10:8,12:10,13:9,14:14,15:15,16:22,17:23,18:24,20:13,23:[1,16],25:17,26:18,27:19,28:[1,20],29:[1,11],30:12,33:26,34:27,35:[1,28],36:[1,30],37:[1,31],38:[1,32],39:[1,33],46:[1,25],47:[1,136],49:21,51:[1,34],52:29},{9:[2,62],40:[2,62],45:[2,62]},{9:[2,61],40:[2,61],45:[2,61]},{9:[1,38],10:48,12:72,13:53,14:14,15:15,16:22,17:23,18:24,20:13,23:[1,16],25:17,26:18,27:19,28:[1,20],29:[1,73],30:12,33:74,34:27,35:[1,28],36:[1,30],37:[1,31],38:[1,32],39:[1,33],44:137,46:[1,25],49:21,51:[1,34],52:29},{9:[1,38],10:48,12:72,13:53,14:14,15:15,16:22,17:23,18:24,20:13,23:[1,16],25:17,26:18,27:19,28:[1,20],29:[1,73],30:12,33:74,34:27,35:[1,28],36:[1,30],37:[1,31],38:[1,32],39:[1,33],40:[2,59],44:124,46:[1,25],49:21,51:[1,34],52:29},{9:[2,65],40:[2,65],45:[2,65]},{9:[2,66],40:[2,66],45:[2,66]},{9:[2,67],40:[2,67],45:[2,67]},{9:[2,68],40:[2,68],45:[2,68]},{5:[2,70],9:[2,70],11:[2,70],19:[2,70],21:[2,70],22:[2,70],23:[2,70],24:[2,70],28:[2,70],29:[2,70],31:[2,70],32:[2,70],35:[2,70],36:[2,70],37:[2,70],38:[2,70],39:[2,70],40:[2,70],45:[2,70],46:[2,70],47:[2,70],48:[2,70],50:[2,70],51:[2,70],53:[2,70],54:[2,70],55:[2,70]},{7:138,9:[1,7]},{5:[2,71],9:[2,71],11:[2,71],19:[2,71],21:[2,71],22:[2,71],23:[2,71],24:[2,71],28:[2,71],29:[2,71],31:[2,71],32:[2,71],35:[2,71],36:[2,71],37:[2,71],38:[2,71],39:[2,71],40:[2,71],45:[2,71],46:[2,71],47:[2,71],48:[2,71],50:[2,71],51:[2,71],53:[2,71],54:[2,71],55:[2,71]},{7:139,9:[1,7]},{5:[2,72],9:[2,72],11:[2,72],19:[2,72],21:[2,72],22:[2,72],23:[2,72],24:[2,72],28:[2,72],29:[2,72],31:[2,72],32:[2,72],35:[2,72],36:[2,72],37:[2,72],38:[2,72],39:[2,72],40:[2,72],45:[2,72],46:[2,72],47:[2,72],48:[2,72],50:[2,72],51:[2,72],53:[2,72],54:[2,72],55:[2,72]},{9:[2,63],40:[2,63],45:[2,63]},{8:77,9:[1,38],10:8,12:10,13:9,14:14,15:15,16:22,17:23,18:24,20:13,23:[1,16],25:17,26:18,27:19,28:[1,20],29:[1,11],30:12,33:26,34:27,35:[1,28],36:[1,30],37:[1,31],38:[1,32],39:[1,33],46:[1,25],47:[1,140],49:21,51:[1,34],52:29},{8:77,9:[1,38],10:8,12:10,13:9,14:14,15:15,16:22,17:23,18:24,20:13,23:[1,16],25:17,26:18,27:19,28:[1,20],29:[1,11],30:12,33:26,34:27,35:[1,28],36:[1,30],37:[1,31],38:[1,32],39:[1,33],46:[1,25],47:[1,141],49:21,51:[1,34],52:29},{5:[2,73],9:[2,73],11:[2,73],19:[2,73],21:[2,73],22:[2,73],23:[2,73],24:[2,73],28:[2,73],29:[2,73],31:[2,73],32:[2,73],35:[2,73],36:[2,73],37:[2,73],38:[2,73],39:[2,73],40:[2,73],45:[2,73],46:[2,73],47:[2,73],48:[2,73],50:[2,73],51:[2,73],53:[2,73],54:[2,73],55:[2,73]},{5:[2,74],9:[2,74],11:[2,74],19:[2,74],21:[2,74],22:[2,74],23:[2,74],24:[2,74],28:[2,74],29:[2,74],31:[2,74],32:[2,74],35:[2,74],36:[2,74],37:[2,74],38:[2,74],39:[2,74],40:[2,74],45:[2,74],46:[2,74],47:[2,74],48:[2,74],50:[2,74],51:[2,74],53:[2,74],54:[2,74],55:[2,74]}],
defaultActions: {3:[2,2],35:[2,1]},
parseError: function parseError(str, hash) {
    if (hash.recoverable) {
        this.trace(str);
    } else {
        throw new Error(str);
    }
},
parse: function parse(input) {
    var self = this, stack = [0], vstack = [null], lstack = [], table = this.table, yytext = '', yylineno = 0, yyleng = 0, recovering = 0, TERROR = 2, EOF = 1;
    var args = lstack.slice.call(arguments, 1);
    this.lexer.setInput(input);
    this.lexer.yy = this.yy;
    this.yy.lexer = this.lexer;
    this.yy.parser = this;
    if (typeof this.lexer.yylloc == 'undefined') {
        this.lexer.yylloc = {};
    }
    var yyloc = this.lexer.yylloc;
    lstack.push(yyloc);
    var ranges = this.lexer.options && this.lexer.options.ranges;
    if (typeof this.yy.parseError === 'function') {
        this.parseError = this.yy.parseError;
    } else {
        this.parseError = Object.getPrototypeOf(this).parseError;
    }
    function popStack(n) {
        stack.length = stack.length - 2 * n;
        vstack.length = vstack.length - n;
        lstack.length = lstack.length - n;
    }
    function lex() {
        var token;
        token = self.lexer.lex() || EOF;
        if (typeof token !== 'number') {
            token = self.symbols_[token] || token;
        }
        return token;
    }
    var symbol, preErrorSymbol, state, action, a, r, yyval = {}, p, len, newState, expected;
    while (true) {
        state = stack[stack.length - 1];
        if (this.defaultActions[state]) {
            action = this.defaultActions[state];
        } else {
            if (symbol === null || typeof symbol == 'undefined') {
                symbol = lex();
            }
            action = table[state] && table[state][symbol];
        }
                    if (typeof action === 'undefined' || !action.length || !action[0]) {
                var errStr = '';
                expected = [];
                for (p in table[state]) {
                    if (this.terminals_[p] && p > TERROR) {
                        expected.push('\'' + this.terminals_[p] + '\'');
                    }
                }
                if (this.lexer.showPosition) {
                    errStr = 'Parse error on line ' + (yylineno + 1) + ':\n' + this.lexer.showPosition() + '\nExpecting ' + expected.join(', ') + ', got \'' + (this.terminals_[symbol] || symbol) + '\'';
                } else {
                    errStr = 'Parse error on line ' + (yylineno + 1) + ': Unexpected ' + (symbol == EOF ? 'end of input' : '\'' + (this.terminals_[symbol] || symbol) + '\'');
                }
                this.parseError(errStr, {
                    text: this.lexer.match,
                    token: this.terminals_[symbol] || symbol,
                    line: this.lexer.yylineno,
                    loc: yyloc,
                    expected: expected
                });
            }
        if (action[0] instanceof Array && action.length > 1) {
            throw new Error('Parse Error: multiple actions possible at state: ' + state + ', token: ' + symbol);
        }
        switch (action[0]) {
        case 1:
            stack.push(symbol);
            vstack.push(this.lexer.yytext);
            lstack.push(this.lexer.yylloc);
            stack.push(action[1]);
            symbol = null;
            if (!preErrorSymbol) {
                yyleng = this.lexer.yyleng;
                yytext = this.lexer.yytext;
                yylineno = this.lexer.yylineno;
                yyloc = this.lexer.yylloc;
                if (recovering > 0) {
                    recovering--;
                }
            } else {
                symbol = preErrorSymbol;
                preErrorSymbol = null;
            }
            break;
        case 2:
            len = this.productions_[action[1]][1];
            yyval.$ = vstack[vstack.length - len];
            yyval._$ = {
                first_line: lstack[lstack.length - (len || 1)].first_line,
                last_line: lstack[lstack.length - 1].last_line,
                first_column: lstack[lstack.length - (len || 1)].first_column,
                last_column: lstack[lstack.length - 1].last_column
            };
            if (ranges) {
                yyval._$.range = [
                    lstack[lstack.length - (len || 1)].range[0],
                    lstack[lstack.length - 1].range[1]
                ];
            }
            r = this.performAction.apply(yyval, [
                yytext,
                yyleng,
                yylineno,
                this.yy,
                action[1],
                vstack,
                lstack
            ].concat(args));
            if (typeof r !== 'undefined') {
                return r;
            }
            if (len) {
                stack = stack.slice(0, -1 * len * 2);
                vstack = vstack.slice(0, -1 * len);
                lstack = lstack.slice(0, -1 * len);
            }
            stack.push(this.productions_[action[1]][0]);
            vstack.push(yyval.$);
            lstack.push(yyval._$);
            newState = table[stack[stack.length - 2]][stack[stack.length - 1]];
            stack.push(newState);
            break;
        case 3:
            return true;
        }
    }
    return true;
}};
/* generated by jison-lex 0.2.1 */
var lexer = (function(){
var lexer = {

EOF:1,

parseError:function parseError(str, hash) {
        if (this.yy.parser) {
            this.yy.parser.parseError(str, hash);
        } else {
            throw new Error(str);
        }
    },

// resets the lexer, sets new input
setInput:function (input) {
        this._input = input;
        this._more = this._backtrack = this.done = false;
        this.yylineno = this.yyleng = 0;
        this.yytext = this.matched = this.match = '';
        this.conditionStack = ['INITIAL'];
        this.yylloc = {
            first_line: 1,
            first_column: 0,
            last_line: 1,
            last_column: 0
        };
        if (this.options.ranges) {
            this.yylloc.range = [0,0];
        }
        this.offset = 0;
        return this;
    },

// consumes and returns one char from the input
input:function () {
        var ch = this._input[0];
        this.yytext += ch;
        this.yyleng++;
        this.offset++;
        this.match += ch;
        this.matched += ch;
        var lines = ch.match(/(?:\r\n?|\n).*/g);
        if (lines) {
            this.yylineno++;
            this.yylloc.last_line++;
        } else {
            this.yylloc.last_column++;
        }
        if (this.options.ranges) {
            this.yylloc.range[1]++;
        }

        this._input = this._input.slice(1);
        return ch;
    },

// unshifts one char (or a string) into the input
unput:function (ch) {
        var len = ch.length;
        var lines = ch.split(/(?:\r\n?|\n)/g);

        this._input = ch + this._input;
        this.yytext = this.yytext.substr(0, this.yytext.length - len - 1);
        //this.yyleng -= len;
        this.offset -= len;
        var oldLines = this.match.split(/(?:\r\n?|\n)/g);
        this.match = this.match.substr(0, this.match.length - 1);
        this.matched = this.matched.substr(0, this.matched.length - 1);

        if (lines.length - 1) {
            this.yylineno -= lines.length - 1;
        }
        var r = this.yylloc.range;

        this.yylloc = {
            first_line: this.yylloc.first_line,
            last_line: this.yylineno + 1,
            first_column: this.yylloc.first_column,
            last_column: lines ?
                (lines.length === oldLines.length ? this.yylloc.first_column : 0)
                 + oldLines[oldLines.length - lines.length].length - lines[0].length :
              this.yylloc.first_column - len
        };

        if (this.options.ranges) {
            this.yylloc.range = [r[0], r[0] + this.yyleng - len];
        }
        this.yyleng = this.yytext.length;
        return this;
    },

// When called from action, caches matched text and appends it on next action
more:function () {
        this._more = true;
        return this;
    },

// When called from action, signals the lexer that this rule fails to match the input, so the next matching rule (regex) should be tested instead.
reject:function () {
        if (this.options.backtrack_lexer) {
            this._backtrack = true;
        } else {
            return this.parseError('Lexical error on line ' + (this.yylineno + 1) + '. You can only invoke reject() in the lexer when the lexer is of the backtracking persuasion (options.backtrack_lexer = true).\n' + this.showPosition(), {
                text: "",
                token: null,
                line: this.yylineno
            });

        }
        return this;
    },

// retain first n characters of the match
less:function (n) {
        this.unput(this.match.slice(n));
    },

// displays already matched input, i.e. for error messages
pastInput:function () {
        var past = this.matched.substr(0, this.matched.length - this.match.length);
        return (past.length > 20 ? '...':'') + past.substr(-20).replace(/\n/g, "");
    },

// displays upcoming input, i.e. for error messages
upcomingInput:function () {
        var next = this.match;
        if (next.length < 20) {
            next += this._input.substr(0, 20-next.length);
        }
        return (next.substr(0,20) + (next.length > 20 ? '...' : '')).replace(/\n/g, "");
    },

// displays the character position where the lexing error occurred, i.e. for error messages
showPosition:function () {
        var pre = this.pastInput();
        var c = new Array(pre.length + 1).join("-");
        return pre + this.upcomingInput() + "\n" + c + "^";
    },

// test the lexed token: return FALSE when not a match, otherwise return token
test_match:function (match, indexed_rule) {
        var token,
            lines,
            backup;

        if (this.options.backtrack_lexer) {
            // save context
            backup = {
                yylineno: this.yylineno,
                yylloc: {
                    first_line: this.yylloc.first_line,
                    last_line: this.last_line,
                    first_column: this.yylloc.first_column,
                    last_column: this.yylloc.last_column
                },
                yytext: this.yytext,
                match: this.match,
                matches: this.matches,
                matched: this.matched,
                yyleng: this.yyleng,
                offset: this.offset,
                _more: this._more,
                _input: this._input,
                yy: this.yy,
                conditionStack: this.conditionStack.slice(0),
                done: this.done
            };
            if (this.options.ranges) {
                backup.yylloc.range = this.yylloc.range.slice(0);
            }
        }

        lines = match[0].match(/(?:\r\n?|\n).*/g);
        if (lines) {
            this.yylineno += lines.length;
        }
        this.yylloc = {
            first_line: this.yylloc.last_line,
            last_line: this.yylineno + 1,
            first_column: this.yylloc.last_column,
            last_column: lines ?
                         lines[lines.length - 1].length - lines[lines.length - 1].match(/\r?\n?/)[0].length :
                         this.yylloc.last_column + match[0].length
        };
        this.yytext += match[0];
        this.match += match[0];
        this.matches = match;
        this.yyleng = this.yytext.length;
        if (this.options.ranges) {
            this.yylloc.range = [this.offset, this.offset += this.yyleng];
        }
        this._more = false;
        this._backtrack = false;
        this._input = this._input.slice(match[0].length);
        this.matched += match[0];
        token = this.performAction.call(this, this.yy, this, indexed_rule, this.conditionStack[this.conditionStack.length - 1]);
        if (this.done && this._input) {
            this.done = false;
        }
        if (token) {
            return token;
        } else if (this._backtrack) {
            // recover context
            for (var k in backup) {
                this[k] = backup[k];
            }
            return false; // rule action called reject() implying the next rule should be tested instead.
        }
        return false;
    },

// return next match in input
next:function () {
        if (this.done) {
            return this.EOF;
        }
        if (!this._input) {
            this.done = true;
        }

        var token,
            match,
            tempMatch,
            index;
        if (!this._more) {
            this.yytext = '';
            this.match = '';
        }
        var rules = this._currentRules();
        for (var i = 0; i < rules.length; i++) {
            tempMatch = this._input.match(this.rules[rules[i]]);
            if (tempMatch && (!match || tempMatch[0].length > match[0].length)) {
                match = tempMatch;
                index = i;
                if (this.options.backtrack_lexer) {
                    token = this.test_match(tempMatch, rules[i]);
                    if (token !== false) {
                        return token;
                    } else if (this._backtrack) {
                        match = false;
                        continue; // rule action called reject() implying a rule MISmatch.
                    } else {
                        // else: this is a lexer rule which consumes input without producing a token (e.g. whitespace)
                        return false;
                    }
                } else if (!this.options.flex) {
                    break;
                }
            }
        }
        if (match) {
            token = this.test_match(match, rules[index]);
            if (token !== false) {
                return token;
            }
            // else: this is a lexer rule which consumes input without producing a token (e.g. whitespace)
            return false;
        }
        if (this._input === "") {
            return this.EOF;
        } else {
            return this.parseError('Lexical error on line ' + (this.yylineno + 1) + '. Unrecognized text.\n' + this.showPosition(), {
                text: "",
                token: null,
                line: this.yylineno
            });
        }
    },

// return next match that has a token
lex:function lex() {
        var r = this.next();
        if (r) {
            return r;
        } else {
            return this.lex();
        }
    },

// activates a new lexer condition state (pushes the new lexer condition state onto the condition stack)
begin:function begin(condition) {
        this.conditionStack.push(condition);
    },

// pop the previously active lexer condition state off the condition stack
popState:function popState() {
        var n = this.conditionStack.length - 1;
        if (n > 0) {
            return this.conditionStack.pop();
        } else {
            return this.conditionStack[0];
        }
    },

// produce the lexer rule set which is active for the currently active lexer condition state
_currentRules:function _currentRules() {
        if (this.conditionStack.length && this.conditionStack[this.conditionStack.length - 1]) {
            return this.conditions[this.conditionStack[this.conditionStack.length - 1]].rules;
        } else {
            return this.conditions["INITIAL"].rules;
        }
    },

// return the currently active lexer condition state; when an index argument is provided it produces the N-th previous condition state, if available
topState:function topState(n) {
        n = this.conditionStack.length - 1 - Math.abs(n || 0);
        if (n >= 0) {
            return this.conditionStack[n];
        } else {
            return "INITIAL";
        }
    },

// alias for begin(condition)
pushState:function pushState(condition) {
        this.begin(condition);
    },

// return the number of states currently on the stack
stateStackSize:function stateStackSize() {
        return this.conditionStack.length;
    },
options: [],
performAction: function anonymous(yy,yy_,$avoiding_name_collisions,YY_START) {

var YYSTATE=YY_START;
switch($avoiding_name_collisions) {
case 0:this.pushState("INSIDE_PARENS"); return "(";
break;
case 1:this.popState(); return ")";
break;
case 2:this.pushState("INITIAL"); return "{";
break;
case 3:this.popState(); return "}";
break;
case 4:this.pushState("INITIAL"); return "[";
break;
case 5:this.popState(); return "]";
break;
case 6:return "|";
break;
case 7:return YY_START === "INITIAL" ? "NEWLINE" : null;
break;
case 8:/* skip other whitespace */
break;
case 9:return YY_START === "INITIAL" ? "NEWLINE" : null; /* skip comments */
break;
case 10:return "SIGN";
break;
case 11:return "AND";
break;
case 12:return "OR";
break;
case 13:return "NEG_NUMBER";
break;
case 14:return "NEG_NUMBER";
break;
case 15:return "NUMBER";
break;
case 16:return "NUMBER";
break;
case 17:return "REGEX";
break;
case 18:return "*";
break;
case 19:return "/";
break;
case 20:return "-";
break;
case 21:return "+";
break;
case 22:return "%";
break;
case 23:return "@";
break;
case 24:return "=";
break;
case 25:return "!";
break;
case 26:return ":";
break;
case 27:return "DOT";
break;
case 28:return ",";
break;
case 29:return "STRING";
break;
case 30:return "STRING";
break;
case 31:return "IDENTIFIER";
break;
case 32:return "JAVASCRIPT";
break;
case 33:return "EOF";
break;
case 34:return "INVALID";
break;
}
},
rules: [/^(?:\()/,/^(?:\))/,/^(?:\{)/,/^(?:\})/,/^(?:\[)/,/^(?:\])/,/^(?:\|)/,/^(?:(\r\n?|\n))/,/^(?:[^\S\r\n]+)/,/^(?:\/\/.*?(\r\n?|\n))/,/^(?:==|!=|<=|>=|<|>)/,/^(?:and)/,/^(?:or)/,/^(?:-[0-9]+)/,/^(?:-([0-9]+)?\.[0-9]+)/,/^(?:[0-9]+)/,/^(?:([0-9]+)?\.[0-9]+)/,/^(?:\/(\\.|[^\/\n])*\/[gim]*)/,/^(?:\*)/,/^(?:\/)/,/^(?:-)/,/^(?:\+)/,/^(?:\%)/,/^(?:@)/,/^(?:=)/,/^(?:\!)/,/^(?:\:)/,/^(?:\.)/,/^(?:,)/,/^(?:\"(\\.|[^"\n])*?\")/,/^(?:\'(\\.|[^'\n])*?\')/,/^(?:[a-zA-Z_$][a-zA-Z0-9_$]*)/,/^(?:```.*?```)/,/^(?:$)/,/^(?:.)/],
conditions: {"INSIDE_PARENS":{"rules":[0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34],"inclusive":true},"INITIAL":{"rules":[0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34],"inclusive":true}}
};
return lexer;
})();
parser.lexer = lexer;
function Parser () {
  this.yy = {};
}
Parser.prototype = parser;parser.Parser = Parser;
return new Parser;
})();

parser.yy = require('./parse-tree.js');
module.exports = parser;
